{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ef454377",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Ruta del archivo GTF\n",
    "ruta_archivo = 'hg38.ncbiRefSeq.gtf'\n",
    "\n",
    "# Columnas que esperamos en el DataFrame\n",
    "columnas = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
    "\n",
    "# Leer el archivo GTF en un DataFrame de pandas\n",
    "df = pd.read_csv(ruta_archivo, sep='\\t', comment='#', header=None, names=columnas)\n",
    "\n",
    "# Ajustar la anchura de la columna para que se muestren todas las columnas\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "# Visualizar el DataFrame completo\n",
    "df = df[df['feature'] == 'transcript'].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3de6f832",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# Definir una función para extraer los valores de cada campo usando expresiones regulares\n",
    "def extraer_valor(cadena, campo):\n",
    "    coincidencia = re.search(f'{campo} \"(.*?)\"', cadena)\n",
    "    if coincidencia:\n",
    "        return coincidencia.group(1)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Aplicar la función a cada fila de la columna 'atributos' para extraer los valores de cada campo\n",
    "df['gene_id'] = df['attributes'].apply(lambda x: extraer_valor(x, 'gene_id'))\n",
    "df['transcript_id'] = df['attributes'].apply(lambda x: extraer_valor(x, 'transcript_id'))\n",
    "df['gene_name'] = df['attributes'].apply(lambda x: extraer_valor(x, 'gene_name'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8ed16222",
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = df['transcript_id'].str.match('.*NM.*|.*NR.*')\n",
    "\n",
    "# Aplicar la máscara al DataFrame para filtrar los valores\n",
    "df_filtrado = df[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4d4c622b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lista de cromosomas\n",
    "cromosomas = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10',\n",
    "              'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19',\n",
    "              'chr20', 'chr21', 'chr22', 'chrX', 'chrY', 'chrM']\n",
    "\n",
    "# Filtrar el DataFrame para mantener solo los cromosomas de interés\n",
    "df_filtrado = df_filtrado[df_filtrado['seqname'].isin(cromosomas)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a2a6e367",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calcular el máximo de la columna 'end' para cada grupo 'gene_id'\n",
    "maximos_por_gene = df_filtrado.groupby(['gene_id','seqname'])['end'].max().reset_index()\n",
    "\n",
    "# Renombrar la columna 'end' como 'pos_max'\n",
    "maximos_por_gene = maximos_por_gene.rename(columns={'end': 'pos_max'})\n",
    "\n",
    "# Calcular el mínimo de la columna 'start' para cada grupo 'gene_id'\n",
    "minimos_por_gene = df_filtrado.groupby(['gene_id','seqname'])['start'].min().reset_index()\n",
    "\n",
    "# Renombrar la columna 'start' como 'pos_min'\n",
    "minimos_por_gene = minimos_por_gene.rename(columns={'start': 'pos_min'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "89d19df2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Suponiendo que tienes los DataFrames df_maximos_por_gene y df_minimos_por_gene que quieres combinar\n",
    "\n",
    "# Combinar los DataFrames a lo ancho por la columna 'gene_id'\n",
    "df_combinado = pd.merge(minimos_por_gene, maximos_por_gene, on=['gene_id','seqname'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ac98f65a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>pos_min</th>\n",
       "      <th>pos_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58345183</td>\n",
       "      <td>58353492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58351970</td>\n",
       "      <td>58355183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>chr10</td>\n",
       "      <td>50799409</td>\n",
       "      <td>50885627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9067708</td>\n",
       "      <td>9116229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9065177</td>\n",
       "      <td>9068055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28770</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52842760</td>\n",
       "      <td>52894998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28771</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52726453</td>\n",
       "      <td>52827336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28772</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>chr7</td>\n",
       "      <td>143381345</td>\n",
       "      <td>143391111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28773</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>4004445</td>\n",
       "      <td>4143030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28774</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>77562416</td>\n",
       "      <td>77683398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28775 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_id seqname    pos_min    pos_max\n",
       "0          A1BG   chr19   58345183   58353492\n",
       "1      A1BG-AS1   chr19   58351970   58355183\n",
       "2          A1CF   chr10   50799409   50885627\n",
       "3           A2M   chr12    9067708    9116229\n",
       "4       A2M-AS1   chr12    9065177    9068055\n",
       "...         ...     ...        ...        ...\n",
       "28770    ZYG11A    chr1   52842760   52894998\n",
       "28771    ZYG11B    chr1   52726453   52827336\n",
       "28772       ZYX    chr7  143381345  143391111\n",
       "28773     ZZEF1   chr17    4004445    4143030\n",
       "28774      ZZZ3    chr1   77562416   77683398\n",
       "\n",
       "[28775 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combinado"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e16c755",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27\n"
     ]
    }
   ],
   "source": [
    "# Contar cuántos duplicados hay en la columna 'gene_id'\n",
    "numero_de_duplicados = df_combinado.duplicated(subset=['gene_id']).sum()\n",
    "\n",
    "# Imprimir el número de duplicados\n",
    "print(numero_de_duplicados)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "17c9c531",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>pos_min</th>\n",
       "      <th>pos_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1591604</td>\n",
       "      <td>1602520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1591604</td>\n",
       "      <td>1602520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1346</th>\n",
       "      <td>ASMT</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1595455</td>\n",
       "      <td>1643081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1347</th>\n",
       "      <td>ASMT</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1595455</td>\n",
       "      <td>1643081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1348</th>\n",
       "      <td>ASMTL</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1403139</td>\n",
       "      <td>1453756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1349</th>\n",
       "      <td>ASMTL</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1403139</td>\n",
       "      <td>1453756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1350</th>\n",
       "      <td>ASMTL-AS1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1400531</td>\n",
       "      <td>1415421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1351</th>\n",
       "      <td>ASMTL-AS1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1400531</td>\n",
       "      <td>1415421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3065</th>\n",
       "      <td>CD99</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2691295</td>\n",
       "      <td>2741309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3066</th>\n",
       "      <td>CD99</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2691295</td>\n",
       "      <td>2741309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3068</th>\n",
       "      <td>CD99P1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2609265</td>\n",
       "      <td>2657229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3069</th>\n",
       "      <td>CD99P1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2609265</td>\n",
       "      <td>2657229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190</th>\n",
       "      <td>CRLF2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1190490</td>\n",
       "      <td>1212649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4191</th>\n",
       "      <td>CRLF2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1190490</td>\n",
       "      <td>1212649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4256</th>\n",
       "      <td>CSF2RA</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1268814</td>\n",
       "      <td>1325218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4257</th>\n",
       "      <td>CSF2RA</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1268814</td>\n",
       "      <td>1325218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4780</th>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>chrX</td>\n",
       "      <td>156025658</td>\n",
       "      <td>156028183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4781</th>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57212178</td>\n",
       "      <td>57214703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4983</th>\n",
       "      <td>DHRSX</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2219506</td>\n",
       "      <td>2500976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4984</th>\n",
       "      <td>DHRSX</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2219506</td>\n",
       "      <td>2500976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8094</th>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>chrX</td>\n",
       "      <td>304759</td>\n",
       "      <td>318796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8095</th>\n",
       "      <td>GTPBP6</td>\n",
       "      <td>chrY</td>\n",
       "      <td>304759</td>\n",
       "      <td>318796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9282</th>\n",
       "      <td>IL3RA</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1336785</td>\n",
       "      <td>1382689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9283</th>\n",
       "      <td>IL3RA</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1336785</td>\n",
       "      <td>1382689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9298</th>\n",
       "      <td>IL9R</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155997696</td>\n",
       "      <td>156010817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9299</th>\n",
       "      <td>IL9R</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57184216</td>\n",
       "      <td>57197337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10610</th>\n",
       "      <td>LINC00102</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2612991</td>\n",
       "      <td>2615347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10611</th>\n",
       "      <td>LINC00102</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2612991</td>\n",
       "      <td>2615347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10612</th>\n",
       "      <td>LINC00106</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1397025</td>\n",
       "      <td>1399412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10613</th>\n",
       "      <td>LINC00106</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1397025</td>\n",
       "      <td>1399412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10913</th>\n",
       "      <td>LINC00685</td>\n",
       "      <td>chrX</td>\n",
       "      <td>319145</td>\n",
       "      <td>321332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10914</th>\n",
       "      <td>LINC00685</td>\n",
       "      <td>chrY</td>\n",
       "      <td>319145</td>\n",
       "      <td>321332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15301</th>\n",
       "      <td>MIR3690</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1293918</td>\n",
       "      <td>1293992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15302</th>\n",
       "      <td>MIR3690</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1293918</td>\n",
       "      <td>1293992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16160</th>\n",
       "      <td>MIR6089</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2609191</td>\n",
       "      <td>2609254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16161</th>\n",
       "      <td>MIR6089</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2609191</td>\n",
       "      <td>2609254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18956</th>\n",
       "      <td>P2RY8</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1462581</td>\n",
       "      <td>1537185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18957</th>\n",
       "      <td>P2RY8</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1462581</td>\n",
       "      <td>1537185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19865</th>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>276356</td>\n",
       "      <td>303356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19866</th>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>276356</td>\n",
       "      <td>303356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>PPP2R3B</td>\n",
       "      <td>chrX</td>\n",
       "      <td>333933</td>\n",
       "      <td>386907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20326</th>\n",
       "      <td>PPP2R3B</td>\n",
       "      <td>chrY</td>\n",
       "      <td>333933</td>\n",
       "      <td>386907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22859</th>\n",
       "      <td>SHOX</td>\n",
       "      <td>chrX</td>\n",
       "      <td>624344</td>\n",
       "      <td>659411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22860</th>\n",
       "      <td>SHOX</td>\n",
       "      <td>chrY</td>\n",
       "      <td>624344</td>\n",
       "      <td>659411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23124</th>\n",
       "      <td>SLC25A6</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1386152</td>\n",
       "      <td>1392113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23125</th>\n",
       "      <td>SLC25A6</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1386152</td>\n",
       "      <td>1392113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24618</th>\n",
       "      <td>SPRY3</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155612586</td>\n",
       "      <td>155782459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24619</th>\n",
       "      <td>SPRY3</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56954316</td>\n",
       "      <td>56968979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27354</th>\n",
       "      <td>VAMP7</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155881345</td>\n",
       "      <td>155943769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27355</th>\n",
       "      <td>VAMP7</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57067865</td>\n",
       "      <td>57130289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27555</th>\n",
       "      <td>WASIR1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>156014564</td>\n",
       "      <td>156016830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27556</th>\n",
       "      <td>WASIR1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57201084</td>\n",
       "      <td>57203350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27851</th>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2486435</td>\n",
       "      <td>2500976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27852</th>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2486435</td>\n",
       "      <td>2500976</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_id seqname    pos_min    pos_max\n",
       "583      AKAP17A    chrX    1591604    1602520\n",
       "584      AKAP17A    chrY    1591604    1602520\n",
       "1346        ASMT    chrX    1595455    1643081\n",
       "1347        ASMT    chrY    1595455    1643081\n",
       "1348       ASMTL    chrX    1403139    1453756\n",
       "1349       ASMTL    chrY    1403139    1453756\n",
       "1350   ASMTL-AS1    chrX    1400531    1415421\n",
       "1351   ASMTL-AS1    chrY    1400531    1415421\n",
       "3065        CD99    chrX    2691295    2741309\n",
       "3066        CD99    chrY    2691295    2741309\n",
       "3068      CD99P1    chrX    2609265    2657229\n",
       "3069      CD99P1    chrY    2609265    2657229\n",
       "4190       CRLF2    chrX    1190490    1212649\n",
       "4191       CRLF2    chrY    1190490    1212649\n",
       "4256      CSF2RA    chrX    1268814    1325218\n",
       "4257      CSF2RA    chrY    1268814    1325218\n",
       "4780    DDX11L16    chrX  156025658  156028183\n",
       "4781    DDX11L16    chrY   57212178   57214703\n",
       "4983       DHRSX    chrX    2219506    2500976\n",
       "4984       DHRSX    chrY    2219506    2500976\n",
       "8094      GTPBP6    chrX     304759     318796\n",
       "8095      GTPBP6    chrY     304759     318796\n",
       "9282       IL3RA    chrX    1336785    1382689\n",
       "9283       IL3RA    chrY    1336785    1382689\n",
       "9298        IL9R    chrX  155997696  156010817\n",
       "9299        IL9R    chrY   57184216   57197337\n",
       "10610  LINC00102    chrX    2612991    2615347\n",
       "10611  LINC00102    chrY    2612991    2615347\n",
       "10612  LINC00106    chrX    1397025    1399412\n",
       "10613  LINC00106    chrY    1397025    1399412\n",
       "10913  LINC00685    chrX     319145     321332\n",
       "10914  LINC00685    chrY     319145     321332\n",
       "15301    MIR3690    chrX    1293918    1293992\n",
       "15302    MIR3690    chrY    1293918    1293992\n",
       "16160    MIR6089    chrX    2609191    2609254\n",
       "16161    MIR6089    chrY    2609191    2609254\n",
       "18956      P2RY8    chrX    1462581    1537185\n",
       "18957      P2RY8    chrY    1462581    1537185\n",
       "19865     PLCXD1    chrX     276356     303356\n",
       "19866     PLCXD1    chrY     276356     303356\n",
       "20325    PPP2R3B    chrX     333933     386907\n",
       "20326    PPP2R3B    chrY     333933     386907\n",
       "22859       SHOX    chrX     624344     659411\n",
       "22860       SHOX    chrY     624344     659411\n",
       "23124    SLC25A6    chrX    1386152    1392113\n",
       "23125    SLC25A6    chrY    1386152    1392113\n",
       "24618      SPRY3    chrX  155612586  155782459\n",
       "24619      SPRY3    chrY   56954316   56968979\n",
       "27354      VAMP7    chrX  155881345  155943769\n",
       "27355      VAMP7    chrY   57067865   57130289\n",
       "27555     WASIR1    chrX  156014564  156016830\n",
       "27556     WASIR1    chrY   57201084   57203350\n",
       "27851      ZBED1    chrX    2486435    2500976\n",
       "27852      ZBED1    chrY    2486435    2500976"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtrar duplicados basados en la columna 'gene_id'\n",
    "duplicados_por_gene_id = df_combinado[df_combinado.duplicated(subset=['gene_id'], keep=False)]\n",
    "\n",
    "# Imprimir el DataFrame con los duplicados por 'gene_id'\n",
    "duplicados_por_gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e5a63e16",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_combinado = df_combinado.drop_duplicates('gene_id')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f32fdcba",
   "metadata": {},
   "source": [
    "# Esta parte solo hay que hacerla una vez para conseguir un excel mejorado con los símbolos de los genes que interactúan entre ellos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "e178a1a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ruta del archivo CSV\n",
    "ruta_archivo_csv = 'genes_id_3.csv'\n",
    "\n",
    "# Cargar el archivo CSV en un DataFrame\n",
    "df_genes_h = pd.read_csv(ruta_archivo_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "4c13617d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define la función para extraer el Official Symbol\n",
    "def extraer_official_symbol(cadena):\n",
    "    coincidencia = re.search(r'Official Symbol: (\\w+-\\w+|\\w+)', cadena)\n",
    "    if coincidencia:\n",
    "        return coincidencia.group(1)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Aplica la función a cada fila de la columna 'descripcion' para extraer el \"Official Symbol\"\n",
    "df_genes_h['Symbol'] = df_genes_h['secuencia'].apply(extraer_official_symbol)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "c0a136a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extraer_symbol_alternativo(cadena):\n",
    "    coincidencia = re.search(r'1. (\\w+)', cadena)\n",
    "    if coincidencia:\n",
    "        return coincidencia.group(1)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Aplicar la función a cada fila de la columna 'descripcion' para extraer el \"Official Symbol\"\n",
    "df_genes_h['Symbol_alt'] = df_genes_h['secuencia'].apply(extraer_symbol_alternativo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "160c8f76",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes_h['Symbol'].fillna(df_genes_h['Symbol_alt'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "162fa52a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extraer_numero_cromosoma(informacion):\n",
    "    coincidencia = re.search(r'Chromosome: (\\w+)', informacion)\n",
    "    if coincidencia:\n",
    "        return coincidencia.group(1)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Aplicar la función a cada fila de la columna 'informacion' para extraer el número del cromosoma\n",
    "df_genes_h['numero_cromosoma'] = df_genes_h['secuencia'].apply(extraer_numero_cromosoma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "f1698b27",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extraer_numero_cromosoma_alt(informacion):\n",
    "    coincidencia = re.search(r'Chromosome (\\w+)', informacion)\n",
    "    if coincidencia:\n",
    "        return coincidencia.group(1)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Aplicar la función a cada fila de la columna 'informacion' para extraer el número del cromosoma\n",
    "df_genes_h['numero_cromosoma_alt'] = df_genes_h['secuencia'].apply(extraer_numero_cromosoma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "6d6719d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes_h['numero_cromosoma'].fillna(df_genes_h['numero_cromosoma_alt'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "fa9bb4ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Especifica la ruta y el nombre del archivo CSV\n",
    "ruta_archivo_csv = 'genes_nuevo.csv'\n",
    "\n",
    "# Convierte el DataFrame a CSV\n",
    "df_genes_h.to_csv(ruta_archivo_csv, index=False)\n",
    "\n",
    "# El argumento index=False evita que se guarde el índice del DataFrame como una columna adicional en el archivo CSV"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "079451e5",
   "metadata": {},
   "source": [
    "# A partir de aquí comienza la segunda parte en la que vamos a cruzar los dos dataframes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "27f56cbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ruta del archivo CSV\n",
    "ruta_archivo_csv = 'Excel_genes_bueno.xlsx'\n",
    "\n",
    "# Cargar el archivo CSV en un DataFrame\n",
    "df_genes = pd.read_excel(ruta_archivo_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4dfe61c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes.drop(['secuencia','Symbol_alt','numero_cromosoma_alt','id_nucleotidos','anotacion'],axis = 1,inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8761ac78",
   "metadata": {},
   "outputs": [],
   "source": [
    "registros_sin_symbol = df_genes[df_genes['Symbol'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fad238e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>numero_cromosoma</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, Symbol, numero_cromosoma]\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "registros_sin_symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "17044cc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "id                  20050\n",
       "Symbol              20050\n",
       "numero_cromosoma    20042\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c29cf2b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Cruzar los DataFrames por los campos especificados\n",
    "df_merged = pd.merge(df_genes, df_combinado, left_on='Symbol', right_on='gene_id', how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e1027902",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged_left = df_merged[df_merged['gene_id'].isna()]\n",
    "df_merged_inner = df_merged[~df_merged['gene_id'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a15fadf2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>numero_cromosoma</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>pos_min</th>\n",
       "      <th>pos_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>171</td>\n",
       "      <td>AFD1</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>206</td>\n",
       "      <td>AK4P1</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>406</td>\n",
       "      <td>BMAL1</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>425</td>\n",
       "      <td>ARVD4</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>574</td>\n",
       "      <td>BAGE</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20014</th>\n",
       "      <td>102723170</td>\n",
       "      <td>IGHV3-64D</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20024</th>\n",
       "      <td>102724334</td>\n",
       "      <td>LOC102724334</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20038</th>\n",
       "      <td>103504728</td>\n",
       "      <td>MIR3687-2</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20041</th>\n",
       "      <td>104355133</td>\n",
       "      <td>ABHD15-AS1</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20045</th>\n",
       "      <td>105376906</td>\n",
       "      <td>ERVK-19</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>699 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id        Symbol numero_cromosoma gene_id seqname  pos_min  \\\n",
       "120          171          AFD1                9     NaN     NaN      NaN   \n",
       "144          206         AK4P1               17     NaN     NaN      NaN   \n",
       "299          406         BMAL1               11     NaN     NaN      NaN   \n",
       "313          425         ARVD4                2     NaN     NaN      NaN   \n",
       "398          574          BAGE               21     NaN     NaN      NaN   \n",
       "...          ...           ...              ...     ...     ...      ...   \n",
       "20014  102723170     IGHV3-64D               14     NaN     NaN      NaN   \n",
       "20024  102724334  LOC102724334               21     NaN     NaN      NaN   \n",
       "20038  103504728     MIR3687-2               21     NaN     NaN      NaN   \n",
       "20041  104355133    ABHD15-AS1               17     NaN     NaN      NaN   \n",
       "20045  105376906       ERVK-19               19     NaN     NaN      NaN   \n",
       "\n",
       "       pos_max  \n",
       "120        NaN  \n",
       "144        NaN  \n",
       "299        NaN  \n",
       "313        NaN  \n",
       "398        NaN  \n",
       "...        ...  \n",
       "20014      NaN  \n",
       "20024      NaN  \n",
       "20038      NaN  \n",
       "20041      NaN  \n",
       "20045      NaN  \n",
       "\n",
       "[699 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged_left"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2324f314",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>numero_cromosoma</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>pos_min</th>\n",
       "      <th>pos_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58345183.0</td>\n",
       "      <td>58353492.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>12</td>\n",
       "      <td>A2M</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9067708.0</td>\n",
       "      <td>9116229.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>8</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>18170467.0</td>\n",
       "      <td>18223689.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>8</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>chr8</td>\n",
       "      <td>18391282.0</td>\n",
       "      <td>18401218.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12</td>\n",
       "      <td>SERPINA3</td>\n",
       "      <td>14</td>\n",
       "      <td>SERPINA3</td>\n",
       "      <td>chr14</td>\n",
       "      <td>94612391.0</td>\n",
       "      <td>94624053.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20044</th>\n",
       "      <td>105375355</td>\n",
       "      <td>UPK3B</td>\n",
       "      <td>7</td>\n",
       "      <td>UPK3B</td>\n",
       "      <td>chr7</td>\n",
       "      <td>76510552.0</td>\n",
       "      <td>76516522.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20046</th>\n",
       "      <td>105377105</td>\n",
       "      <td>FLNB-AS1</td>\n",
       "      <td>3</td>\n",
       "      <td>FLNB-AS1</td>\n",
       "      <td>chr3</td>\n",
       "      <td>58162547.0</td>\n",
       "      <td>58170635.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20047</th>\n",
       "      <td>105378803</td>\n",
       "      <td>LRRC53</td>\n",
       "      <td>1</td>\n",
       "      <td>LRRC53</td>\n",
       "      <td>chr1</td>\n",
       "      <td>74469376.0</td>\n",
       "      <td>74512611.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20048</th>\n",
       "      <td>107126285</td>\n",
       "      <td>SEPTIN14P20</td>\n",
       "      <td>20</td>\n",
       "      <td>SEPTIN14P20</td>\n",
       "      <td>chr20</td>\n",
       "      <td>64290385.0</td>\n",
       "      <td>64303354.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20049</th>\n",
       "      <td>107282092</td>\n",
       "      <td>BOLA2-SMG1P6</td>\n",
       "      <td>16</td>\n",
       "      <td>BOLA2-SMG1P6</td>\n",
       "      <td>chr16</td>\n",
       "      <td>29442905.0</td>\n",
       "      <td>29454964.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19351 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id        Symbol numero_cromosoma       gene_id seqname  \\\n",
       "0              1          A1BG               19          A1BG   chr19   \n",
       "1              2           A2M               12           A2M   chr12   \n",
       "2              9          NAT1                8          NAT1    chr8   \n",
       "3             10          NAT2                8          NAT2    chr8   \n",
       "4             12      SERPINA3               14      SERPINA3   chr14   \n",
       "...          ...           ...              ...           ...     ...   \n",
       "20044  105375355         UPK3B                7         UPK3B    chr7   \n",
       "20046  105377105      FLNB-AS1                3      FLNB-AS1    chr3   \n",
       "20047  105378803        LRRC53                1        LRRC53    chr1   \n",
       "20048  107126285   SEPTIN14P20               20   SEPTIN14P20   chr20   \n",
       "20049  107282092  BOLA2-SMG1P6               16  BOLA2-SMG1P6   chr16   \n",
       "\n",
       "          pos_min     pos_max  \n",
       "0      58345183.0  58353492.0  \n",
       "1       9067708.0   9116229.0  \n",
       "2      18170467.0  18223689.0  \n",
       "3      18391282.0  18401218.0  \n",
       "4      94612391.0  94624053.0  \n",
       "...           ...         ...  \n",
       "20044  76510552.0  76516522.0  \n",
       "20046  58162547.0  58170635.0  \n",
       "20047  74469376.0  74512611.0  \n",
       "20048  64290385.0  64303354.0  \n",
       "20049  29442905.0  29454964.0  \n",
       "\n",
       "[19351 rows x 7 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged_inner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "eefe9f4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\mojoy\\AppData\\Local\\Temp\\ipykernel_6368\\2470853460.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_merged_inner['pos_min'] = df_merged_inner['pos_min'].astype(int)\n",
      "C:\\Users\\mojoy\\AppData\\Local\\Temp\\ipykernel_6368\\2470853460.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_merged_inner['pos_max'] = df_merged_inner['pos_max'].astype(int)\n"
     ]
    }
   ],
   "source": [
    "df_merged_inner['pos_min'] = df_merged_inner['pos_min'].astype(int)\n",
    "df_merged_inner['pos_max'] = df_merged_inner['pos_max'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "06e25299",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Creo que estos genes que no están en la base de datos del GTF a la fecha en la que se hizo su recogida. La fecha de la\\nbase de datos es de 2022 y por ejemplo para los genes AK4P1 y BMAL1 sí que tienen una secuencia de nucleótidos asociada pero que\\nse ha hecho en 2023.'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''Creo que estos genes que no están en la base de datos del GTF a la fecha en la que se hizo su recogida. La fecha de la\n",
    "base de datos es de 2022 y por ejemplo para los genes AK4P1 y BMAL1 sí que tienen una secuencia de nucleótidos asociada pero que\n",
    "se ha hecho en 2023.'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c13894b",
   "metadata": {},
   "source": [
    "# Vamos a ver si el problema persiste para las otras bases de datos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "id": "4bd88af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Ruta del archivo GTF\n",
    "ruta_archivo = 'hg38.refGene.gtf'\n",
    "\n",
    "# Columnas que esperamos en el DataFrame\n",
    "columnas = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
    "\n",
    "# Leer el archivo GTF en un DataFrame de pandas\n",
    "df = pd.read_csv(ruta_archivo, sep='\\t', comment='#', header=None, names=columnas)\n",
    "\n",
    "# Ajustar la anchura de la columna para que se muestren todas las columnas\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "# Visualizar el DataFrame completo\n",
    "df = df[df['feature'] == 'transcript'].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "id": "2c6c9576",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# Definir una función para extraer los valores de cada campo usando expresiones regulares\n",
    "def extraer_valor(cadena, campo):\n",
    "    coincidencia = re.search(f'{campo} \"(.*?)\"', cadena)\n",
    "    if coincidencia:\n",
    "        return coincidencia.group(1)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Aplicar la función a cada fila de la columna 'atributos' para extraer los valores de cada campo\n",
    "df['gene_id'] = df['attributes'].apply(lambda x: extraer_valor(x, 'gene_id'))\n",
    "df['transcript_id'] = df['attributes'].apply(lambda x: extraer_valor(x, 'transcript_id'))\n",
    "df['gene_name'] = df['attributes'].apply(lambda x: extraer_valor(x, 'gene_name'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "bcd1d417",
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = df['transcript_id'].str.match('.*NM.*|.*NR.*')\n",
    "\n",
    "# Aplicar la máscara al DataFrame para filtrar los valores\n",
    "df_filtrado = df[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "id": "a976c532",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lista de cromosomas\n",
    "cromosomas = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10',\n",
    "              'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19',\n",
    "              'chr20', 'chr21', 'chr22', 'chrX', 'chrY', 'chrM']\n",
    "\n",
    "# Filtrar el DataFrame para mantener solo los cromosomas de interés\n",
    "df_filtrado = df_filtrado[df_filtrado['seqname'].isin(cromosomas)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "id": "1b28d8db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calcular el máximo de la columna 'end' para cada grupo 'gene_id'\n",
    "maximos_por_gene = df_filtrado.groupby(['gene_id','seqname'])['end'].max().reset_index()\n",
    "\n",
    "# Renombrar la columna 'end' como 'pos_max'\n",
    "maximos_por_gene = maximos_por_gene.rename(columns={'end': 'pos_max'})\n",
    "\n",
    "# Calcular el mínimo de la columna 'start' para cada grupo 'gene_id'\n",
    "minimos_por_gene = df_filtrado.groupby(['gene_id','seqname'])['start'].min().reset_index()\n",
    "\n",
    "# Renombrar la columna 'start' como 'pos_min'\n",
    "minimos_por_gene = minimos_por_gene.rename(columns={'start': 'pos_min'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "id": "c53b86c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Suponiendo que tienes los DataFrames df_maximos_por_gene y df_minimos_por_gene que quieres combinar\n",
    "\n",
    "# Combinar los DataFrames a lo ancho por la columna 'gene_id'\n",
    "df_combinado = pd.merge(minimos_por_gene, maximos_por_gene, on=['gene_id','seqname'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "id": "d658c07b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>pos_min</th>\n",
       "      <th>pos_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58345183</td>\n",
       "      <td>58353492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58351970</td>\n",
       "      <td>58355183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>chr10</td>\n",
       "      <td>50799409</td>\n",
       "      <td>50885675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9067708</td>\n",
       "      <td>9116229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>chr12</td>\n",
       "      <td>9065177</td>\n",
       "      <td>9068055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28398</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52842760</td>\n",
       "      <td>52894995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28399</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>chr1</td>\n",
       "      <td>52726453</td>\n",
       "      <td>52827336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28400</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>chr7</td>\n",
       "      <td>143381345</td>\n",
       "      <td>143391111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28401</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>4004445</td>\n",
       "      <td>4143030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28402</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>77562416</td>\n",
       "      <td>77683427</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28403 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_id seqname    pos_min    pos_max\n",
       "0          A1BG   chr19   58345183   58353492\n",
       "1      A1BG-AS1   chr19   58351970   58355183\n",
       "2          A1CF   chr10   50799409   50885675\n",
       "3           A2M   chr12    9067708    9116229\n",
       "4       A2M-AS1   chr12    9065177    9068055\n",
       "...         ...     ...        ...        ...\n",
       "28398    ZYG11A    chr1   52842760   52894995\n",
       "28399    ZYG11B    chr1   52726453   52827336\n",
       "28400       ZYX    chr7  143381345  143391111\n",
       "28401     ZZEF1   chr17    4004445    4143030\n",
       "28402      ZZZ3    chr1   77562416   77683427\n",
       "\n",
       "[28403 rows x 4 columns]"
      ]
     },
     "execution_count": 295,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combinado"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "id": "5d92991e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "153\n"
     ]
    }
   ],
   "source": [
    "# Contar cuántos duplicados hay en la columna 'gene_id'\n",
    "numero_de_duplicados = df_combinado.duplicated(subset=['gene_id']).sum()\n",
    "\n",
    "# Imprimir el número de duplicados\n",
    "print(numero_de_duplicados)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "id": "e0b6c85f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>pos_min</th>\n",
       "      <th>pos_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1591604</td>\n",
       "      <td>1602518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>AKAP17A</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1591604</td>\n",
       "      <td>1602518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1316</th>\n",
       "      <td>ASMT</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1595455</td>\n",
       "      <td>1643081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1317</th>\n",
       "      <td>ASMT</td>\n",
       "      <td>chrY</td>\n",
       "      <td>1595455</td>\n",
       "      <td>1643081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1318</th>\n",
       "      <td>ASMTL</td>\n",
       "      <td>chrX</td>\n",
       "      <td>1403139</td>\n",
       "      <td>1453762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27213</th>\n",
       "      <td>WASIR1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57201084</td>\n",
       "      <td>57203350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27394</th>\n",
       "      <td>XGY2</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2752296</td>\n",
       "      <td>2774996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27395</th>\n",
       "      <td>XGY2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2752296</td>\n",
       "      <td>2774996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27504</th>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>2486414</td>\n",
       "      <td>2500976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27505</th>\n",
       "      <td>ZBED1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2486414</td>\n",
       "      <td>2500976</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>245 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id seqname   pos_min   pos_max\n",
       "568    AKAP17A    chrX   1591604   1602518\n",
       "569    AKAP17A    chrY   1591604   1602518\n",
       "1316      ASMT    chrX   1595455   1643081\n",
       "1317      ASMT    chrY   1595455   1643081\n",
       "1318     ASMTL    chrX   1403139   1453762\n",
       "...        ...     ...       ...       ...\n",
       "27213   WASIR1    chrY  57201084  57203350\n",
       "27394     XGY2    chrX   2752296   2774996\n",
       "27395     XGY2    chrY   2752296   2774996\n",
       "27504    ZBED1    chrX   2486414   2500976\n",
       "27505    ZBED1    chrY   2486414   2500976\n",
       "\n",
       "[245 rows x 4 columns]"
      ]
     },
     "execution_count": 297,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtrar duplicados basados en la columna 'gene_id'\n",
    "duplicados_por_gene_id = df_combinado[df_combinado.duplicated(subset=['gene_id'], keep=False)]\n",
    "\n",
    "# Imprimir el DataFrame con los duplicados por 'gene_id'\n",
    "duplicados_por_gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "d9e00d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_combinado = df_combinado.drop_duplicates('gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "05e79584",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ruta del archivo CSV\n",
    "ruta_archivo_csv = 'Excel_genes_bueno.xlsx'\n",
    "\n",
    "# Cargar el archivo CSV en un DataFrame\n",
    "df_genes = pd.read_excel(ruta_archivo_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "4cab08b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes.drop(['secuencia','Symbol_alt','numero_cromosoma_alt','id_nucleotidos','anotacion'],axis = 1,inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "id": "a0b5a7cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Cruzar los DataFrames por los campos especificados\n",
    "df_merged = pd.merge(df_genes, df_combinado, left_on='Symbol', right_on='gene_id', how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "id": "bb10f98f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged_left = df_merged[df_merged['gene_id'].isna()]\n",
    "df_merged_inner = df_merged[~df_merged['gene_id'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "id": "bb08a2e6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>numero_cromosoma</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>pos_min</th>\n",
       "      <th>pos_max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>171</td>\n",
       "      <td>AFD1</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>206</td>\n",
       "      <td>AK4P1</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>406</td>\n",
       "      <td>BMAL1</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>425</td>\n",
       "      <td>ARVD4</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>802</td>\n",
       "      <td>CALM1P1</td>\n",
       "      <td>X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20032</th>\n",
       "      <td>102724652</td>\n",
       "      <td>LOC102724652</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20038</th>\n",
       "      <td>103504728</td>\n",
       "      <td>MIR3687-2</td>\n",
       "      <td>21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20041</th>\n",
       "      <td>104355133</td>\n",
       "      <td>ABHD15-AS1</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20045</th>\n",
       "      <td>105376906</td>\n",
       "      <td>ERVK-19</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20048</th>\n",
       "      <td>107126285</td>\n",
       "      <td>SEPTIN14P20</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>877 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id        Symbol numero_cromosoma gene_id seqname  pos_min  \\\n",
       "120          171          AFD1                9     NaN     NaN      NaN   \n",
       "144          206         AK4P1               17     NaN     NaN      NaN   \n",
       "299          406         BMAL1               11     NaN     NaN      NaN   \n",
       "313          425         ARVD4                2     NaN     NaN      NaN   \n",
       "568          802       CALM1P1                X     NaN     NaN      NaN   \n",
       "...          ...           ...              ...     ...     ...      ...   \n",
       "20032  102724652  LOC102724652               21     NaN     NaN      NaN   \n",
       "20038  103504728     MIR3687-2               21     NaN     NaN      NaN   \n",
       "20041  104355133    ABHD15-AS1               17     NaN     NaN      NaN   \n",
       "20045  105376906       ERVK-19               19     NaN     NaN      NaN   \n",
       "20048  107126285   SEPTIN14P20               20     NaN     NaN      NaN   \n",
       "\n",
       "       pos_max  \n",
       "120        NaN  \n",
       "144        NaN  \n",
       "299        NaN  \n",
       "313        NaN  \n",
       "568        NaN  \n",
       "...        ...  \n",
       "20032      NaN  \n",
       "20038      NaN  \n",
       "20041      NaN  \n",
       "20045      NaN  \n",
       "20048      NaN  \n",
       "\n",
       "[877 rows x 7 columns]"
      ]
     },
     "execution_count": 303,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged_left"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "a4ef2b36",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_malos = df[df['gene_id'] == 'BMAL1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "e7d9f1da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Con esta base de datos obtenemos más registros que no nos cruzan.'"
      ]
     },
     "execution_count": 304,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''Con esta base de datos obtenemos más registros que no nos cruzan.'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc3b231e",
   "metadata": {},
   "source": [
    "# Vamos a comprobar la última"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "95563b27",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Ruta del archivo GTF\n",
    "ruta_archivo = 'hg38.knownGene.gtf'\n",
    "\n",
    "# Columnas que esperamos en el DataFrame\n",
    "columnas = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
    "\n",
    "# Leer el archivo GTF en un DataFrame de pandas\n",
    "df = pd.read_csv(ruta_archivo, sep='\\t', comment='#', header=None, names=columnas)\n",
    "\n",
    "# Ajustar la anchura de la columna para que se muestren todas las columnas\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "# Visualizar el DataFrame completo\n",
    "df = df[df['feature'] == 'transcript'].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "283e867c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# Definir una función para extraer los valores de cada campo usando expresiones regulares\n",
    "def extraer_valor(cadena, campo):\n",
    "    coincidencia = re.search(f'{campo} \"(.*?)\"', cadena)\n",
    "    if coincidencia:\n",
    "        return coincidencia.group(1)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Aplicar la función a cada fila de la columna 'atributos' para extraer los valores de cada campo\n",
    "df['gene_id'] = df['attributes'].apply(lambda x: extraer_valor(x, 'gene_id'))\n",
    "df['transcript_id'] = df['attributes'].apply(lambda x: extraer_valor(x, 'transcript_id'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "id": "e0598b1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lista de cromosomas\n",
    "cromosomas = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10',\n",
    "              'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19',\n",
    "              'chr20', 'chr21', 'chr22', 'chrX', 'chrY', 'chrM']\n",
    "\n",
    "# Filtrar el DataFrame para mantener solo los cromosomas de interés\n",
    "df_filtrado = df[df['seqname'].isin(cromosomas)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "f6d93fb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000456328.2\"; transcript_id \"ENST00000456328.2\";</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000619216.1\"; transcript_id \"ENST00000619216.1\";</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000473358.1\"; transcript_id \"ENST00000473358.1\";</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30267</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000469289.1\"; transcript_id \"ENST00000469289.1\";</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000607096.1\"; transcript_id \"ENST00000607096.1\";</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "      <td>ENST00000607096.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274026</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>58534</td>\n",
       "      <td>60100</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000643054.1\"; transcript_id \"ENST00000643054.1\";</td>\n",
       "      <td>ENST00000643054.1</td>\n",
       "      <td>ENST00000643054.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274027</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>82985</td>\n",
       "      <td>95067</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000647527.1\"; transcript_id \"ENST00000647527.1\";</td>\n",
       "      <td>ENST00000647527.1</td>\n",
       "      <td>ENST00000647527.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274028</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>98305</td>\n",
       "      <td>102177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000644360.1\"; transcript_id \"ENST00000644360.1\";</td>\n",
       "      <td>ENST00000644360.1</td>\n",
       "      <td>ENST00000644360.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274029</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>117629</td>\n",
       "      <td>119279</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000645951.1\"; transcript_id \"ENST00000645951.1\";</td>\n",
       "      <td>ENST00000645951.1</td>\n",
       "      <td>ENST00000645951.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274030</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>knownGene</td>\n",
       "      <td>transcript</td>\n",
       "      <td>198642</td>\n",
       "      <td>199281</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENST00000646767.1\"; transcript_id \"ENST00000646767.1\";</td>\n",
       "      <td>ENST00000646767.1</td>\n",
       "      <td>ENST00000646767.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>274031 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    seqname     source     feature   start     end score  \\\n",
       "0                      chr1  knownGene  transcript   11869   14409     .   \n",
       "1                      chr1  knownGene  transcript   17369   17436     .   \n",
       "2                      chr1  knownGene  transcript   29554   31097     .   \n",
       "3                      chr1  knownGene  transcript   30267   31109     .   \n",
       "4                      chr1  knownGene  transcript   30366   30503     .   \n",
       "...                     ...        ...         ...     ...     ...   ...   \n",
       "274026  chrY_KZ208924v1_fix  knownGene  transcript   58534   60100     .   \n",
       "274027  chrY_KZ208924v1_fix  knownGene  transcript   82985   95067     .   \n",
       "274028  chrY_KZ208924v1_fix  knownGene  transcript   98305  102177     .   \n",
       "274029  chrY_KZ208924v1_fix  knownGene  transcript  117629  119279     .   \n",
       "274030  chrY_KZ208924v1_fix  knownGene  transcript  198642  199281     .   \n",
       "\n",
       "       strand frame  \\\n",
       "0           +     .   \n",
       "1           -     .   \n",
       "2           +     .   \n",
       "3           +     .   \n",
       "4           +     .   \n",
       "...       ...   ...   \n",
       "274026      -     .   \n",
       "274027      +     .   \n",
       "274028      -     .   \n",
       "274029      +     .   \n",
       "274030      -     .   \n",
       "\n",
       "                                                              attributes  \\\n",
       "0       gene_id \"ENST00000456328.2\"; transcript_id \"ENST00000456328.2\";    \n",
       "1       gene_id \"ENST00000619216.1\"; transcript_id \"ENST00000619216.1\";    \n",
       "2       gene_id \"ENST00000473358.1\"; transcript_id \"ENST00000473358.1\";    \n",
       "3       gene_id \"ENST00000469289.1\"; transcript_id \"ENST00000469289.1\";    \n",
       "4       gene_id \"ENST00000607096.1\"; transcript_id \"ENST00000607096.1\";    \n",
       "...                                                                  ...   \n",
       "274026  gene_id \"ENST00000643054.1\"; transcript_id \"ENST00000643054.1\";    \n",
       "274027  gene_id \"ENST00000647527.1\"; transcript_id \"ENST00000647527.1\";    \n",
       "274028  gene_id \"ENST00000644360.1\"; transcript_id \"ENST00000644360.1\";    \n",
       "274029  gene_id \"ENST00000645951.1\"; transcript_id \"ENST00000645951.1\";    \n",
       "274030  gene_id \"ENST00000646767.1\"; transcript_id \"ENST00000646767.1\";    \n",
       "\n",
       "                  gene_id      transcript_id  \n",
       "0       ENST00000456328.2  ENST00000456328.2  \n",
       "1       ENST00000619216.1  ENST00000619216.1  \n",
       "2       ENST00000473358.1  ENST00000473358.1  \n",
       "3       ENST00000469289.1  ENST00000469289.1  \n",
       "4       ENST00000607096.1  ENST00000607096.1  \n",
       "...                   ...                ...  \n",
       "274026  ENST00000643054.1  ENST00000643054.1  \n",
       "274027  ENST00000647527.1  ENST00000647527.1  \n",
       "274028  ENST00000644360.1  ENST00000644360.1  \n",
       "274029  ENST00000645951.1  ENST00000645951.1  \n",
       "274030  ENST00000646767.1  ENST00000646767.1  \n",
       "\n",
       "[274031 rows x 11 columns]"
      ]
     },
     "execution_count": 288,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "457ed2da",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Con esta base de datos no tenemos un campo por el que cruzar con la nuestra.'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7c7069d",
   "metadata": {},
   "source": [
    "# Cruzamos con nuestra tabla de interacciones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fda8eb96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Vamos a cargar la tabla de interacciones'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''Vamos a cargar la tabla de interacciones'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bffcfdd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importa las bibliotecas necesarias\n",
    "from pyspark.sql import SparkSession\n",
    "from pyspark.sql.functions import col, desc, count, countDistinct, explode, split, size\n",
    "# Crea una instancia de SparkSession\n",
    "spark = SparkSession.builder.appName(\"Ejemplo\").getOrCreate()\n",
    "\n",
    "path_4 = 'interactions\\interactions.txt'\n",
    "df_int = spark.read.csv(path_4, header=True, inferSchema=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4b984534",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_int_homo = df_int.where( (col('#tax_id') == 9606) & (col('tax_id') == 9606))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fbae0c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_int_homo_filtrado = df_int_homo.select('gene_id','interactant_id','pubmed_id_list')\n",
    "df_int_homo_filtrado = df_int_homo_filtrado.drop_duplicates()\n",
    "df_int_homo_filtrado = df_int_homo_filtrado.toPandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "60301c06",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_int_homo_filtrado = df_int_homo_filtrado.sort_values(by='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ed587bde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>interactant_id</th>\n",
       "      <th>pubmed_id_list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2886</td>\n",
       "      <td>21900206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772216</th>\n",
       "      <td>1</td>\n",
       "      <td>80854</td>\n",
       "      <td>21988832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>770099</th>\n",
       "      <td>1</td>\n",
       "      <td>84236</td>\n",
       "      <td>33961781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>770098</th>\n",
       "      <td>1</td>\n",
       "      <td>148581</td>\n",
       "      <td>33961781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>769570</th>\n",
       "      <td>1</td>\n",
       "      <td>5655</td>\n",
       "      <td>33961781</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_id interactant_id pubmed_id_list\n",
       "0             1           2886       21900206\n",
       "772216        1          80854       21988832\n",
       "770099        1          84236       33961781\n",
       "770098        1         148581       33961781\n",
       "769570        1           5655       33961781"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_int_homo_filtrado.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "eae81eca",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_1 = pd.merge(df_int_homo_filtrado, df_merged_inner[['id', 'Symbol', 'seqname', 'pos_min', 'pos_max']], left_on='gene_id', right_on='id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "35eb2788",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_1_inner = df_1[~df_1['seqname'].isnull()]\n",
    "\n",
    "# Estos registros que no cruzan como ya he comentado antes, es porque la secuencia se ha incorporado más tarde a la base de datos.\n",
    "df_1_left = df_1[df_1['seqname'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b0ea1b35",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_comprobacion = pd.merge(df_1_left, df_merged_left[['id', 'Symbol', 'seqname', 'pos_min', 'pos_max']], left_on='gene_id', right_on='id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "00a01c69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>interactant_id</th>\n",
       "      <th>pubmed_id_list</th>\n",
       "      <th>id_x</th>\n",
       "      <th>Symbol_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>pos_min_x</th>\n",
       "      <th>pos_max_x</th>\n",
       "      <th>id_y</th>\n",
       "      <th>Symbol_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>pos_min_y</th>\n",
       "      <th>pos_max_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>171</td>\n",
       "      <td>3265</td>\n",
       "      <td>30442766</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>171</td>\n",
       "      <td>AFD1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>206</td>\n",
       "      <td>10212</td>\n",
       "      <td>32393512</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>206</td>\n",
       "      <td>AK4P1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>406</td>\n",
       "      <td>3159</td>\n",
       "      <td>30021884</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>406</td>\n",
       "      <td>BMAL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>425</td>\n",
       "      <td>10618</td>\n",
       "      <td>35696571</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>425</td>\n",
       "      <td>ARVD4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>574</td>\n",
       "      <td>3191</td>\n",
       "      <td>28611215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>574</td>\n",
       "      <td>BAGE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4737</th>\n",
       "      <td>653319</td>\n",
       "      <td>728116</td>\n",
       "      <td>32296183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>653319</td>\n",
       "      <td>MATCAP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4738</th>\n",
       "      <td>653440</td>\n",
       "      <td>100287171</td>\n",
       "      <td>33961781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>653440</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4739</th>\n",
       "      <td>653639</td>\n",
       "      <td>728655</td>\n",
       "      <td>32572027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>653639</td>\n",
       "      <td>LYPLA2P1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4740</th>\n",
       "      <td>100132285</td>\n",
       "      <td>100132285</td>\n",
       "      <td>11172020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100132285</td>\n",
       "      <td>KIR2DS2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4741</th>\n",
       "      <td>100505705</td>\n",
       "      <td>100505705</td>\n",
       "      <td>21255413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100505705</td>\n",
       "      <td>SDIM1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>239 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_id interactant_id pubmed_id_list  id_x Symbol_x seqname_x  \\\n",
       "0           171           3265       30442766   NaN      NaN       NaN   \n",
       "15          206          10212       32393512   NaN      NaN       NaN   \n",
       "16          406           3159       30021884   NaN      NaN       NaN   \n",
       "120         425          10618       35696571   NaN      NaN       NaN   \n",
       "121         574           3191       28611215   NaN      NaN       NaN   \n",
       "...         ...            ...            ...   ...      ...       ...   \n",
       "4737     653319         728116       32296183   NaN      NaN       NaN   \n",
       "4738     653440      100287171       33961781   NaN      NaN       NaN   \n",
       "4739     653639         728655       32572027   NaN      NaN       NaN   \n",
       "4740  100132285      100132285       11172020   NaN      NaN       NaN   \n",
       "4741  100505705      100505705       21255413   NaN      NaN       NaN   \n",
       "\n",
       "      pos_min_x  pos_max_x       id_y  Symbol_y seqname_y  pos_min_y  \\\n",
       "0           NaN        NaN        171      AFD1       NaN        NaN   \n",
       "15          NaN        NaN        206     AK4P1       NaN        NaN   \n",
       "16          NaN        NaN        406     BMAL1       NaN        NaN   \n",
       "120         NaN        NaN        425     ARVD4       NaN        NaN   \n",
       "121         NaN        NaN        574      BAGE       NaN        NaN   \n",
       "...         ...        ...        ...       ...       ...        ...   \n",
       "4737        NaN        NaN     653319   MATCAP1       NaN        NaN   \n",
       "4738        NaN        NaN     653440    WASH6P       NaN        NaN   \n",
       "4739        NaN        NaN     653639  LYPLA2P1       NaN        NaN   \n",
       "4740        NaN        NaN  100132285   KIR2DS2       NaN        NaN   \n",
       "4741        NaN        NaN  100505705     SDIM1       NaN        NaN   \n",
       "\n",
       "      pos_max_y  \n",
       "0           NaN  \n",
       "15          NaN  \n",
       "16          NaN  \n",
       "120         NaN  \n",
       "121         NaN  \n",
       "...         ...  \n",
       "4737        NaN  \n",
       "4738        NaN  \n",
       "4739        NaN  \n",
       "4740        NaN  \n",
       "4741        NaN  \n",
       "\n",
       "[239 rows x 13 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comprobacion.drop_duplicates('gene_id')\n",
    "\n",
    "# Comprobamos que los id son los que ya sabíamos que no tenían secuencia asociada."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "6e570f8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nos quitamos las interacciones donde no se indica el otro gen\n",
    "df_1_inner = df_1_inner[~(df_1_inner['interactant_id'] == '-')]\n",
    "df_1_inner['interactant_id'] = df_1_inner['interactant_id'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6248dd0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ahora vamos a cruzar el df_1_inner con la base de datos de las posiciones para obtener la secuencia del gen que interactúa\n",
    "df_2 = pd.merge(df_1_inner, df_merged_inner[['id', 'Symbol', 'seqname', 'pos_min', 'pos_max']], left_on='interactant_id', right_on='id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "bd535380",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_2_inner = df_2[~df_2['seqname_y'].isnull()]\n",
    "df_2_left = df_2[df_2['seqname_y'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "0f7b72a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>interactant_id</th>\n",
       "      <th>pubmed_id_list</th>\n",
       "      <th>id_x</th>\n",
       "      <th>Symbol_x</th>\n",
       "      <th>seqname_x</th>\n",
       "      <th>pos_min_x</th>\n",
       "      <th>pos_max_x</th>\n",
       "      <th>id_y</th>\n",
       "      <th>Symbol_y</th>\n",
       "      <th>seqname_y</th>\n",
       "      <th>pos_min_y</th>\n",
       "      <th>pos_max_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2886</td>\n",
       "      <td>21900206</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58345183.0</td>\n",
       "      <td>58353492.0</td>\n",
       "      <td>2886.0</td>\n",
       "      <td>GRB7</td>\n",
       "      <td>chr17</td>\n",
       "      <td>39737938.0</td>\n",
       "      <td>39747285.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>80854</td>\n",
       "      <td>21988832</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58345183.0</td>\n",
       "      <td>58353492.0</td>\n",
       "      <td>80854.0</td>\n",
       "      <td>SETD7</td>\n",
       "      <td>chr4</td>\n",
       "      <td>139492974.0</td>\n",
       "      <td>139556219.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>84236</td>\n",
       "      <td>33961781</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58345183.0</td>\n",
       "      <td>58353492.0</td>\n",
       "      <td>84236.0</td>\n",
       "      <td>RHBDD1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>226800159.0</td>\n",
       "      <td>226999210.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>148581</td>\n",
       "      <td>33961781</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58345183.0</td>\n",
       "      <td>58353492.0</td>\n",
       "      <td>148581.0</td>\n",
       "      <td>UBE2U</td>\n",
       "      <td>chr1</td>\n",
       "      <td>64203623.0</td>\n",
       "      <td>64267368.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>5655</td>\n",
       "      <td>33961781</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>chr19</td>\n",
       "      <td>58345183.0</td>\n",
       "      <td>58353492.0</td>\n",
       "      <td>5655.0</td>\n",
       "      <td>KLK10</td>\n",
       "      <td>chr19</td>\n",
       "      <td>51012739.0</td>\n",
       "      <td>51020097.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866147</th>\n",
       "      <td>100506658</td>\n",
       "      <td>100506658</td>\n",
       "      <td>10026224</td>\n",
       "      <td>100506658.0</td>\n",
       "      <td>OCLN</td>\n",
       "      <td>chr5</td>\n",
       "      <td>69492547.0</td>\n",
       "      <td>69558104.0</td>\n",
       "      <td>100506658.0</td>\n",
       "      <td>OCLN</td>\n",
       "      <td>chr5</td>\n",
       "      <td>69492547.0</td>\n",
       "      <td>69558104.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866148</th>\n",
       "      <td>100506742</td>\n",
       "      <td>100506742</td>\n",
       "      <td>11278723,12097332</td>\n",
       "      <td>100506742.0</td>\n",
       "      <td>CASP12</td>\n",
       "      <td>chr11</td>\n",
       "      <td>104885718.0</td>\n",
       "      <td>104898460.0</td>\n",
       "      <td>100506742.0</td>\n",
       "      <td>CASP12</td>\n",
       "      <td>chr11</td>\n",
       "      <td>104885718.0</td>\n",
       "      <td>104898460.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866149</th>\n",
       "      <td>100652748</td>\n",
       "      <td>100652748</td>\n",
       "      <td>30033366</td>\n",
       "      <td>100652748.0</td>\n",
       "      <td>TIMM23B</td>\n",
       "      <td>chr10</td>\n",
       "      <td>49942053.0</td>\n",
       "      <td>49974850.0</td>\n",
       "      <td>100652748.0</td>\n",
       "      <td>TIMM23B</td>\n",
       "      <td>chr10</td>\n",
       "      <td>49942053.0</td>\n",
       "      <td>49974850.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866150</th>\n",
       "      <td>102723631</td>\n",
       "      <td>102723631</td>\n",
       "      <td>32296183</td>\n",
       "      <td>102723631.0</td>\n",
       "      <td>CT45A10</td>\n",
       "      <td>chrX</td>\n",
       "      <td>135881065.0</td>\n",
       "      <td>135893444.0</td>\n",
       "      <td>102723631.0</td>\n",
       "      <td>CT45A10</td>\n",
       "      <td>chrX</td>\n",
       "      <td>135881065.0</td>\n",
       "      <td>135893444.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866151</th>\n",
       "      <td>102724652</td>\n",
       "      <td>102724652</td>\n",
       "      <td>32296183</td>\n",
       "      <td>102724652.0</td>\n",
       "      <td>LOC102724652</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6499203.0</td>\n",
       "      <td>6564489.0</td>\n",
       "      <td>102724652.0</td>\n",
       "      <td>LOC102724652</td>\n",
       "      <td>chr21</td>\n",
       "      <td>6499203.0</td>\n",
       "      <td>6564489.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>858382 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_id  interactant_id     pubmed_id_list         id_x  \\\n",
       "0               1            2886           21900206          1.0   \n",
       "1               1           80854           21988832          1.0   \n",
       "2               1           84236           33961781          1.0   \n",
       "3               1          148581           33961781          1.0   \n",
       "4               1            5655           33961781          1.0   \n",
       "...           ...             ...                ...          ...   \n",
       "866147  100506658       100506658           10026224  100506658.0   \n",
       "866148  100506742       100506742  11278723,12097332  100506742.0   \n",
       "866149  100652748       100652748           30033366  100652748.0   \n",
       "866150  102723631       102723631           32296183  102723631.0   \n",
       "866151  102724652       102724652           32296183  102724652.0   \n",
       "\n",
       "            Symbol_x seqname_x    pos_min_x    pos_max_x         id_y  \\\n",
       "0               A1BG     chr19   58345183.0   58353492.0       2886.0   \n",
       "1               A1BG     chr19   58345183.0   58353492.0      80854.0   \n",
       "2               A1BG     chr19   58345183.0   58353492.0      84236.0   \n",
       "3               A1BG     chr19   58345183.0   58353492.0     148581.0   \n",
       "4               A1BG     chr19   58345183.0   58353492.0       5655.0   \n",
       "...              ...       ...          ...          ...          ...   \n",
       "866147          OCLN      chr5   69492547.0   69558104.0  100506658.0   \n",
       "866148        CASP12     chr11  104885718.0  104898460.0  100506742.0   \n",
       "866149       TIMM23B     chr10   49942053.0   49974850.0  100652748.0   \n",
       "866150       CT45A10      chrX  135881065.0  135893444.0  102723631.0   \n",
       "866151  LOC102724652     chr21    6499203.0    6564489.0  102724652.0   \n",
       "\n",
       "            Symbol_y seqname_y    pos_min_y    pos_max_y  \n",
       "0               GRB7     chr17   39737938.0   39747285.0  \n",
       "1              SETD7      chr4  139492974.0  139556219.0  \n",
       "2             RHBDD1      chr2  226800159.0  226999210.0  \n",
       "3              UBE2U      chr1   64203623.0   64267368.0  \n",
       "4              KLK10     chr19   51012739.0   51020097.0  \n",
       "...              ...       ...          ...          ...  \n",
       "866147          OCLN      chr5   69492547.0   69558104.0  \n",
       "866148        CASP12     chr11  104885718.0  104898460.0  \n",
       "866149       TIMM23B     chr10   49942053.0   49974850.0  \n",
       "866150       CT45A10      chrX  135881065.0  135893444.0  \n",
       "866151  LOC102724652     chr21    6499203.0    6564489.0  \n",
       "\n",
       "[858382 rows x 13 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2_inner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "c2ac7e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "def leer_secuencias_fasta(ruta_archivo):\n",
    "    secuencias = {}  # Diccionario para almacenar las secuencias de cada cromosoma\n",
    "    with gzip.open(ruta_archivo, 'rt') as f:\n",
    "        nombre_cromosoma = ''\n",
    "        secuencia_actual = []\n",
    "        for linea in f:\n",
    "            if linea.startswith('>'):\n",
    "                # Procesar la secuencia anterior y almacenarla si hay una secuencia actual\n",
    "                if nombre_cromosoma and secuencia_actual:\n",
    "                    secuencias[nombre_cromosoma] = ''.join(secuencia_actual)\n",
    "                    secuencia_actual = []  # Reiniciar la lista para la próxima secuencia\n",
    "                nombre_cromosoma = linea.strip()[1:]  # Obtener el nombre del cromosoma\n",
    "            else:\n",
    "                # Almacenar la línea de secuencia actual\n",
    "                secuencia_actual.append(linea.strip())\n",
    "        \n",
    "        # Almacenar la última secuencia si existe\n",
    "        if nombre_cromosoma and secuencia_actual:\n",
    "            secuencias[nombre_cromosoma] = ''.join(secuencia_actual)\n",
    "            \n",
    "    return secuencias\n",
    "\n",
    "# Lista de nombres de archivos FASTA comprimidos por cromosoma\n",
    "file_paths = [\n",
    "    'archivos_fasta_cromosomas/chrM.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chrX.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chrY.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr22.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr21.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr20.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr19.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr18.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr17.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr16.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr15.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr14.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr13.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr12.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr11.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr10.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr9.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr8.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr7.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr6.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr5.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr4.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr3.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr2.fa.gz',\n",
    "    'archivos_fasta_cromosomas/chr1.fa.gz'\n",
    "]\n",
    "\n",
    "# Diccionario para almacenar las secuencias de los cromosomas\n",
    "chromosome_sequences = {}\n",
    "\n",
    "# Leer las secuencias de cada archivo y almacenarlas en el diccionario\n",
    "for file_path in file_paths:\n",
    "    chromosome_name = file_path.split('/')[-1].split('.')[0]\n",
    "    sequences = leer_secuencias_fasta(file_path)\n",
    "    chromosome_sequences[chromosome_name] = sequences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1c2d6793",
   "metadata": {},
   "outputs": [],
   "source": [
    "def obtener_secuencia(inicio, fin, cromosoma):\n",
    "    secuencia = chromosome_sequences[cromosoma][cromosoma][inicio-1:fin]\n",
    "    return secuencia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "28680acb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\mojoy\\AppData\\Local\\Temp\\ipykernel_6368\\2470863266.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_2_inner['pos_min_x'] = df_2_inner['pos_min_x'].astype(int)\n",
      "C:\\Users\\mojoy\\AppData\\Local\\Temp\\ipykernel_6368\\2470863266.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_2_inner['pos_max_x'] = df_2_inner['pos_max_x'].astype(int)\n",
      "C:\\Users\\mojoy\\AppData\\Local\\Temp\\ipykernel_6368\\2470863266.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_2_inner['pos_min_y'] = df_2_inner['pos_min_y'].astype(int)\n",
      "C:\\Users\\mojoy\\AppData\\Local\\Temp\\ipykernel_6368\\2470863266.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_2_inner['pos_max_y'] = df_2_inner['pos_max_y'].astype(int)\n"
     ]
    }
   ],
   "source": [
    "df_2_inner['pos_min_x'] = df_2_inner['pos_min_x'].astype(int)\n",
    "df_2_inner['pos_max_x'] = df_2_inner['pos_max_x'].astype(int)\n",
    "df_2_inner['pos_min_y'] = df_2_inner['pos_min_y'].astype(int)\n",
    "df_2_inner['pos_max_y'] = df_2_inner['pos_max_y'].astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "59bd36fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Crear un diccionario vacío para almacenar los resultados\n",
    "resultados_diccionario_1 = {}\n",
    "\n",
    "# Iterar sobre las filas del DataFrame\n",
    "for index, row in df_2_inner.iterrows():\n",
    "    # Obtener el valor de la columna 'id'\n",
    "    id_valor = row['gene_id']\n",
    "    # Calcular el resultado de la función para la fila actual\n",
    "    resultado = obtener_secuencia(row['pos_min_x'], row['pos_max_x'], row['seqname_x']).upper()\n",
    "    # Almacenar el resultado en el diccionario usando el valor de 'id' como clave\n",
    "    resultados_diccionario_1[id_valor] = resultado\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c874e271",
   "metadata": {},
   "outputs": [],
   "source": [
    "resultados_diccionario_2 = {}\n",
    "\n",
    "# Iterar sobre las filas del DataFrame\n",
    "for index, row in df_2_inner.iterrows():\n",
    "    # Obtener el valor de la columna 'id'\n",
    "    id_valor = row['interactant_id']\n",
    "    # Calcular el resultado de la función para la fila actual\n",
    "    resultado = obtener_secuencia(row['pos_min_y'], row['pos_max_y'], row['seqname_y']).upper()\n",
    "    # Almacenar el resultado en el diccionario usando el valor de 'id' como clave\n",
    "    resultados_diccionario_2[id_valor] = resultado\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "94b48382",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AAACAACCCAATATTTTAAAGTGTGCAAAATATATAAACAGACATTTCATCAAAGATGCTATATAGATGGCAACTAAGCACTTGGGGAAAAAATGCTTAACATCATCAGACATTAAGGAAACGCAAGGAAAACCACTATTAGATATGTCTACATACCTATTAGAATGGCTAAAATAAACTTTTTAAAAAACTAAGCTGGGGCTGGGCGTGGTGGCTCACCCCTGTAATCCCAGCACTTTGGGAGGCTGAGGCGGGTGGATCACTTGAGCTCAGGAGTTGGAGACCAGCCTGGCCACCCAACAAGGTGAAACCCCATCTCTACTAAAAACATAAAAATTAGCTGGGTGTGGTGGTGGGTGCCTGTCAGCCCAGCTACTCAGGAGGCTGAGGCACGAGAATCACTTGAACCCAGGAGGTAGAGGTTGCTGTGAGCCAAGATCACGCCATTGCACTCCAGCCTGGGCAACAAGAGCAAAATTCTGTCTCAAAAACAAATAAACAAGAAAAGGTTTTAAAAATTGATAAACCACTAGCAAGAAAGAGAGAGAGGACACATCAGGCTTGAGAGAAGTGACATCATCATAGACTGTGCAGACATTAAAAGGCATAGGCATATTAGGAACTAGATATTACTAAGGCAGTAAACCAGATCACTAGCAAATTCGTTGAAAAACACAATTCATCAACTATTGGTTGATAATACTGATACGCATCTTCTGTGTGCCAGGGCCTTGAGGCCCTGCTCCAGGCAGAAACTCGGCAGTGGTTGGGAAGGCATCAGATGTCAGATGGCACAAGAGGACTCAGAGCTGAGGGATAATGGGAAAGAACACAGAGGAATCCAGCCATTTCCACAGCGTCCAGCTCTGCTGTGGGAGGCTGGGAACAGCCCAGCACTACCACCCTGGACTGGGAGGACAAGACCACAAAATGCAGCTTCCCTGAACCTCCTCTTGGTGATGGGGTTGATGTGGTCGAGGTAGAGGTGAGTATGTCTGGAGCCTCAAGCAACTGCCCATACCTGCTGGGACCAGGCCGAGGTGCCCAGGGAGTAAGAGGCAGCATCCTGGAAGAGCACAGATGAAAGAGGCCCTGAGAAGTGTTGGTTGGCCAGGTGCTGTGGCTCACACCTGTAATCCCAACACTTTGGGAGGCTGAGGCGGGAGGATCACTTGAGCCCAAGAGTTCAAGACCAGCCTGGGCAACATAGTGAGACTCATCTCTACAAAAAATAAATAATAGAAAGAAAAATGTGAGTTGGCCAGGCATGGTGGCTCATGCCTGTAGTCCCAGCTACTCAGGAGGCTGAGGTGGGAGGATCACTTAAGGCCAGGAGTTCAAGACAAGCTTGGGCAACACAGTGAGACCCTGTCTACAAAAAATAATAATTAGCCAGACGTGGCGATGCATGCCCAGGCTCCCAGCTACTTGGGAGGCTGAGGCAGGAGGATCGCTTGAGCCTGGGAGGTCAAAACTGCAGTGAGCCGGACTGCAACACTGCACTCCAGCCTGGGTGACAGTGTGAGACCCTGTCTCAAAAAAGAAAAAAGAAAAGAAAACTGTGCTCTTAAGAGCCAGTTCTCCACTCCTCTACCTCAGGAGCCACCCCAGAACCCATCCACTTTGAGGACACGAGATCCCAGCCCACTCAGCCCTGGGAGTCCAAAGACATTTTAAACAGAGCCTCTCTTCACATTTATTAATTCCTGGGAGGAATGAGGGAGGCTTCTCCAGCCCCCCAGAGACCCCGGCCTTGTGCTGCAACAGGAGGGGAGGGAGCCAGTCCAGAATCCCCGGCACTTCTGAGGACACCAACAGCACCCTGGGCCCGCGGCTGCATCAGCTTTCTAGACAACGGGAGAAAAGAGAAATGGTGGAGGAGGGGAAATCCTATGCCCTCCTCCTCGCGGAAATCAAGGCCGACCTCCCTGACGCCCCCCCGGAAGGAAGCGCGTGGTCGGCTGCCAGCCGAGACCCCCATCTGCGCCTCCAGCTGCAGGGCCTTTGCTGGGCGGTTCCTCGCCCGCAATTCCCCCACGAGCCCTGGGGAGACCCAGCGCTAACCAGGGGTGCCCAAGGGAGGAGCCGGGAGAGGCCCTCCTGGAGGTGGGCACAGCCCAGGCAAACATCAGCAGACGGAACCAGCACCCGGGACCCAGGGAAACGTCACCTGCCACCAGGAGCTCCACAGGGTCGCTGAGCTCCGATTCGAAGGTGTGGGGCACCCAGGAGCGGTAGCGGCACCTGTAGTTGCCGGCGTGCTGGGGCCCCACGAAGATCAGCTCGAGGTTCGCCGCGGCCCCGGGGGTGCGGACCGTCTTCACGGCCTTCGTCTCGCCCTCGCGCAGCAGCTCGAAGGTGACGTCGGGGATGGGTCCCTCGCAGCGCAGGACGGCATCTCGGCCCGCCAGGACCGCCCCACTCCACGTCGCCCGGAGCTGAGGCCTGGGAGGGGGTCCTGGGCGGAGCGGGCGGGTGGTCGGGCCAGGCCACGCCCCAGGCCACGCCCCAGGCCACACCCCAGGCCACACCCCAGGCCGCGCCCGCGCCTGCGCCTCAGCCCCGGCTTCATCTTCCCTGTCTTGTTCCTGGGCGCAGAGGGCTCCTCCGGCGCCGCCGAGCTGCGCCCGCTGCCTTCACACCCCGGACATCCGCGCCACACGTGAGTGTCCCCGCACCCCTCCCCCAGGACCGTGGAAGTCCCAGATTCGGGCCTGCTCTGGGGCTTCTGGGGTCCCCAGCCTCGGTCACTCTGACGGGTGAAGGGAAGAAGAAATGATGGAGAAAACGGTGCTCCTTCATGGAGGGACCAAAAAAGGACATACAATCTATGTCAATAGATGAGGGTTTTCCCTCCATAAGTTTAACTTTAAAAGTGTATTTAAGGTGCCGGGCGCAGTGGCTCACACCTGTAATCTCAGCACTTTGAGAGGCCGAGGCAGGCGGATTGCCTGAGCTCAGGAGTTCCAGACCAGCCTGGGTAACACGGTGAAACCCCGTCTCTACTAAAATACAAAAAATTAGCCGGGCGTGGCAGCGTGCGCCTGTAATCCCAGCTACTCGGGAGGTGGAGGCAGGAAAATTGCTTGAACCTGGGAGGCAGAGGTTTCAGTGAGCCAAGATTGCGCCACTGCACTCCAGCCCGGGTGACAGAGTGACACTCCGTCTCAAAAACAAACAAACAAAACAAAACTGTAAGGAATCAAATGTTACAAAATCCTGGAAGATTTTGGTAATGCTGTTGATAATTTCAGGGAAATTAATGAAGTACATGTAAGTGCTTGATTTAAATGTGTATTTTATTCAATAGTTTGAGTATCTTGCCATGGAAGATACTAGCCCAGCCTAGCAGAAAAGTGCAATATGTATAATATACTTTCGCATTTTAAAAGTTATATTCCCTATCCATTTTGAAATGCTAGCCAAACTACATGAAGTTATTTGCAATTAATTGTAAGCTAATCAGGAATTTTGAAAGCTTGATTGGATTAACAAACTATAATGTTGAAATTTGGTAAAATCATAATGTTAATTTTTACTGTGAAACGTTTTCTTGTAAGATAGATATACACATCCTAGTTTAATTTTGTGTCTTAGTGTGGATTTACAAATTTACTACAGTTATTCTGAGCCCAGGAACCCAATCAGATTTCAGGCCAGTTTGATACTGGCTGTTCTTAATTCTAACAGGAGAGTAGGACATCATACTAAATGTTATGTCAGTGGGACTGTGTTGTCTTTGGAGCTTAGTAAATTAATCATTTTCTTCAGACTTTAAGGAGAGTGATAAATAAAATTTGGAGTCATAGGATATAGATGTACAATTTAAAGATTAAACATTTTTTTTTTTTTTGAGATGGAGTTTTGCTCTTGTTGCCCAGGCTGGAATGCAGTGGCATGACCTGGACTCACTGCAAACTCCACCTCCCGGGTTAAAGCAATTCTCCTGCCTCAGCCTCCCAAGTGGCTGGGATTACAGGCATGTGCCACCACAGCTGGCTAATTTTGTATTTTTAGTAGAGACAGGGTTTCTCCATGTTGGCCAGGCTGCTCTTGAACTCCTGATTTTGTGATCCACCCGCCTTGGCCTCCCAAAGTGCTGTGATTACAGGCGTGAGCCACCTTGCCTGGCCAAGATTAAACATTTTTCATCAATTGTGGATGGTGGCTTATAACACTGCAAGAATAGGCTGGGCATGGTGACTCACAAGTGTAATCCCAGCACTTTGGGAGGCAGAGGCGGGCAGATCACAAGGTCAGGAGGTCAAGACCAGCCCAGCCAACATAGCAAAACCCGGTCTCTACTAAAAATACGAAAATTAGGTGGGCATGGTGGCATGCGCCTCTAATACCAGCTACTCAGGAGGCTGAGGCAGGAGAACTGCCTGAACCCAGAAGGCGAAAGTTGCAGTGAGCTGAGACTGCACCACTGCACTCCGGCCTGCAAGAATAAAAGTAAGTTATCCAAAAAAAAAACAAAAAGGAGAAGAAAAGGGAGGAATGAAGGAGAGAGATGGGGTTGGGGGGCAGAGATGAGAGAGAGAGAAAGAAGAAAAAAGAAGAAAGAAAAGAGAGAAAGAAGGAAGGAAGAAAGAAAGAAGGAAAGAGAAAGAAAGAAAGAGGAAGGAAGGCAGGAAAGAAAAGAAAGAAAGTGAAAGAAAAGAGAAAACGGTGCTCCCAGCAAAGGGTACGACCTCAAGATCTGGAGGTCAGACCTGTGCCCTCACGCTCTGCTGGGCTCTGAGTCCTGGAGACGTCGATGAGGAAATCAGAGGGACAAACAGAGCTGGGGACGGCGGGGATGGGGAGAAGGACACATCAGCCGTGCGCCTGCGGGACCGGGGCCGGGGCGGGGGGTCCAGTATGACGACTGACATGGCCAGTAATAAAACAAGGCAGCCACAGAGCTGCAGGGATCTGAACAAGAAAACCTGGGGCCAGCCTCCCTGCTGCCGACCACCACGCTACACGTGACCTACGGGCTTGCGTCCGTGGGAAATAAGCCGAAATTCCAGGCGTCCCCAGACCCTCGACCACCGATCGCCTGCTCCCCGCCGGGGTAGGCGATGGGGGCTGAACCAAGGCCCAGAGCGCCGCCGTCGGACGCCCAAGGAAAGAGGGGAAAGACAGGAGGCCCCGAGGGGCACTGAACCCGCGCCCGCCCTCGCTGGTGCCCCGCCAGCTCACCGTCCACGTGCAGCTCCAAGCGCTCGCTGGGCGCGGAGCCCCCGAAAGGCGGCTTCAGGTCCACGTAGACGCAGCTGTAGTTGGCGGAGTCAGCCACGGAAATGTTGTGCAGCTCGAAGAGCGCCTCGGTCCCAGCGGGGCTCTGGAAACGGTGCACGCGGCGCCCGCCCCTGTCCTCGCGCACCAGGGCGAAGCGCGCGCCCTCCAGGGGCGCCAGGCACCGCAGCCGCAAGGCCCTGCCGGACTCCGGCTCCGGGGAGAACTCCGGCGCGGGCAGCGTCTCTGCGGAGCAGCACACGGGCTGACCCGGGCGCCCAGCGGCTGGCTCGGCCCTAACGCGTGGTCTGGCCTCGCGCTCTTTGCCTTGGCTGAGTCTTCTGCCCGCAATGACCTGGCCAGCTTCCTCCCTCGCCTCCTCCGGGCCCGCCCCACCTTTGCATCCCAGTCCTTGGATATCTCACACGCACCCTTTCACTGTTTCTTAGTAATTTGTTTGTCTACAGACACAGACGGGAAGCGGGGACGGGGACATGTTTTGTGCATGGCTTGATGCCTGGTGCACGATGCCGCTCAGTAGTGGATTTGCTCCACGAGCGTCTTCTGAGCACCGCCCCAATGCCCGCCCCACGGAGGGGCCCCTGTGGTCCGCTGGGCATGGGGTTCGCTGTCCGGTGCTCAGAGTTTGCTAAATACCATCAGGTCCAAGGCAAGCGAGCGGAGGAGCTGCCTCAGCAGAGGCCCTCAGTGCTCACCTAGACCTGGCAGCTGGGATGGGACCTGCCAGGTGGTGCCCCCTGTGGCCTGGGCCAGACCTGTTCACCCAGTTATTCCCTGTGGGGCCTGAAGGTTGTCACGCTCATCATCATATTTGCAGTTTAGGACCCTGAATCGGCGGGTGGGCGGATAAAGTTGGTATTGGACCCTGGCCCAGAGCACTGCACTTCCCCAGACTCTACACCTGGTACTGCTCCTGCTCCCCAGGGACCCAGCCGCGTCCCTGTCCCTGCTGGCCCCGGCTCACCATCGCTCAGAATCAGCTCGACCGGCGCGCTGTCCCCGGACCAGCCGTTTTGGTTGTCATGCAGCCGGTAGCGGCAGGTGTAGTGACCTCCATCCCCCAGGGCCACCGCGTTCAGGTGAAAGAAGATGCGATCTGGGCTGGTGCTGCTCCTGGGTACCAGCAGCTCTTTCTCCCCGCGCCGTAGCTGGAAGTCCACTCCACTCAGGGGAGCCACGCAGGTGAGGGTCACCTTGTTGCCAGGGTGCAGGACCTGGGAGGACTCTCCATGGTGCATCAGCACAGGCGGTGGTGGTGCAGCTGCAATGCAGGCAGCATTACTAGGCTGCCCCATCCCTGGAGCTGCCCTCTGCCCAGTCCCCAGACCTCTGCCCTCCGGGTGGCAATCCCAGGAACACCCTTCCATTCTTTTTTTTTTTTTTTTTTGAAACGGAGTTTCGCTCTTGTTGCCCAGGCTGGAGTGCAGTGGCGCCACTCAGCTCACTGCAAGCTCTGCCTCCCAGGTTCAAGGGATTCTCCTGCCTCAGCCTCCTGAGTAGCTGGTACTACAGGCACCTGCCACCACACCTGGCTAATTTTTGTATTTTTAGTAGAGACGGGGTTTCGTCATGTTGGGTAGACTGGTCTCGATCTCTTGACCTCATGATCCGCTGGCCTCAGCCTCCCAAAGTGCTGGGATTATAGGCGTGAGCCACCGCACCCGGCCTCACCCTTCCATTCTTTGGGCATCCTCTGCCCAAACTCCTTTGCAACCATCAGACCCAGGCAACCGTGTGGCCAGCTTCCTGGACCTGGTCCTGGGCAAGGACATTCAGCATGGAGGTCAGGATGAATGTGGTCATGCCCCCTGTATCCTCCTCCCCCAGCAGCCCCCAGAGATGGTTCCCACAGTCACCGAGCTCCTCAATGGTCACAGTAGCGCTGGGCTCAGAGAGGGCGCCTTCCCCATCGGTCCGGTAGCTGCAGCTGTAGTTGCCAGGCTGATGGACTGGAAAGGTGGCCTCCACATCCTCCTGGGCCTCAGGCACCTCCAGAAACTCATGGTCGCCCTCCCGCCTCAGCAGAAAAGTCACACCCCGCAGCACACCTCGGCACACTGCTGTTGTTTTCAGGCCGGGGGTGATCCAGGACACTGGCGCCATCGAGAGCCAGGGAGCAGGCAAGGACTCTGTGGGTGGGGTGGAGTTGAAGAGTGCTTCTGCATAACAGGAGACGTGGGAAGCCCAGCAGAACCTGGGCCCCACACTCATAAGACTGTGAAGGAGACAGGGATGGAGGGCATTGCTGGGAAAAGGCAGGCAGGCACCAGTGGAAACAAGGCATACGGCAAGAGAAAAAGAGTGTGTGGGAGCTGTACGGCATGCCGGGCCTGCTGGGGACACAACCATGTATTCCTTACAACACAGACATCATCCCCATTTGACAGATGGGAAGATTGGGGCTTGGGGCTCAGAGACATCGGGTGACTTGGAGGAAGGGAGACCCCCCAGTCAACAACCTACCCACTGCCTCCTGGCCCCCAATTCATGCCCCGGCTCACTTGGCCCTGTCAGCTCCAGGAGCTTGCTCAGCTGGGTCCATCCTGTGGACAAGCCCGAGCGGCAGCGGTAGCGGCCCTGGGTGTCACCCGTCAGCAGGAACTGGTGCTTGATGGCAGGTGAGTCAAGGTGCACAGGCTCCTGGGCCACCCCATTCTTGAACAGCTGGAAGTCTGGAGTCTCCAGGTGGGCCTGGCACGTCAGCGTCACATTGGCCAAGGGTTTCAGCAGTGATTCGGACTCTGCCCACAGGCTGGGCTGCGTCTCATAAACTGCAAGGGGTGGCTGGGATCAGCTCAGTGCCACAGAGACAGGGCTCCCCCCCGGCCTGGGCCCACCATTCCCAGACCTCACCCCTGCACTCACATATGGCTGCTTCTGTCACTGGGCCCCAGGTGACACCTGCGGAGACAGCCCCCGTAAGGCTCCTGTTCCCGCCCCCTCCCGCCCCAGGGACCCAGACCCCAGGAGGCCTCACCCCACAGCAAGAGAAAGACCACGAGCATGGACATGATGGTCGCGCTCACTCCGGTGCAGTGAGTGTCTGGGGTGAGCGTCTGCAGCAAT'"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resultados_diccionario_1[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "4c85b5cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGTTTCCTTGGGCCTGGAATCTGGACACACAGGGCTCCCCCCCGCCTCTGACTTCTCTGTCCGAAGTCGGGACACCCTCCTACCACCTGTAGAGAAGCGGGAGTGGATCTGAAATAAAATCCAGGAATCTGGGGGTTCCTAGACGGAGCCAGACTTCGGAACGGGTGTCCTGCTACTCCTGCTGGGGCTCCTCCAGGTAAGGGGACCGAGATCGGTCTCAGACGACAGCCTGCCAGGCAACCCCCAGCCGCACCAGGGCCCCAGGCTGTGCCAGTCTCCGGCTGAACTGGCCGGCGGCCCAGGTGGGAGGATTGGAGCCACCTGGCTCTCCCCCTCCCCCGCTCGGTGATGTCAGGCCAGGGGCGGGGGCTGGAGGGGGACAGGTAGTTAAGGGCCTGGCGTCTCCCTCCCTGAAGACGTGGTCCCAGCCGGGTGTCCTGACGCTCGGGGTTCAGGTAAGAACCCAAGCGGGCATTTATCCTTTGGGGGCAGGCACGGCCGGTCTGGAGGCCGGTAGGGAAAGGCTGGGTCTCCTGATGGGTTGCTTGGCCTCCCACTTTTCTCTTTTCTGTCCATTTCCCACCCTGCCCCCTCCTCCCCACCCGCCCTCTTTTCGTCCCCATCCTGGTTTCCAGCGCTCCTTGTTCCCTCCCTGTACACCTCCCTGCAGGCAGCAGCTGAGAATGGGAAGGAAGACCTGATTCTCTCCTCCGACTTGGCTTTCCCGCTGAAACTAGCCCCTCCCCGCACACCCGGGGAGAGCCGCCCTCACTCTCTGGCTCGCCCAGTTTACCCCAGTTTGGGTTTCCCCAAGTCTCTAAGACAGACGTCTCTGCGGGCTGCGGGGAGATGTGGGGAGGGCCCCCTCCACTTTGGAGGGCAGTGAAGGAGAGGGATCCTCTAAATTGTCGAGGCTTCATCTCTCCAGATTGTATGCCCTTCTCAGCAACACCGCCTCCGGCCCTCCGATGGGAAAGTGGAGGCCGGGTGAGACGGGGTCCCTCCTGGCTTCAGCATGAAATGGGCAGGAAGGCTGGAGGGGTGTATGAGGGGCTCTGCAGTGCGGTGGGGCCTGGGGAGAGTGGGGTGGGGGATAGCAGATGGTTTCTTCTGCTTGATGAGGCTAGGGAGAGACCCAGGTTCCCGGTCTGGGTCTCCCTGGCAGATTGAGGACAGAGAATGAATGCATCTGCCGGCAAGGTGGAACGCCCCCCACCCCGCCCCCCAACTAGCCGCACCGCCCCCCAGGAATGCCCTCTTTTGGTCCGCCGGTCTTGGACTCTGCTCTCCCCTCTCCTGATATCCTACTGTATGCTTGGTGGGGGGACCTCATGGAGGCGGGCAGCCTGTAATGATTAACTGCAGAGACACTCCAGCATTCTTCCCGAATTAAACTTTAAAAGAGCTGCCCCGCCTTCTCCCATTGGGTATTTTTCCTGAGTGAGAGTGGGCGGGGCCTCAGAGTCCTGGGTCCTCCCTTATACTAGGAGTCATCCAGGGGTTTCTCACCTTCTCACCTGGGCCCTGCACCCTAAAAAATTATAAGCCTCTCCCGTGGGGGCGGGGCGTGGAGTACTGGGCGGACGGGGGCTCTCCCGCCACATCCAGGCCCCTGGCCCAGGCCTCAGTCCCGCAAGCCCTTCCCCCTGCTCTGGGAAGCAGAGTTGTTTTCTGTTGGCCCTGAACACCGGGCCCCGCCACCTGCCCTTAGGGCCAGCCAGGTGGGGCGTGATTCAGGAAGTAAACAAGGAGGTTGGGAGGACGGGAACGCGGGTGGGGCTGGAACCACTGGGGGAGGAGGCAGCTGGGAAGCCACGGACGGAGTTACCTCGTTGAGGGGACAAACTGATTTTGGGGGTCGGTGTTTCCAACCGAAGCTTTACCGGGGCCTCAGTTTGAAAACTCCAGATCTCTAAGCTGCCAGGCCCACAGCCAGGACCCCTTCCTCCTCCCTGCTCACCCCATTGGTCTTGACGATTCCACCCGCCTCCAGCCTTAGAAGCTTAAGGGCACGGCTGGGGGGCTCTGAAGGCAAACCCCAGCCTTGGACTGGCCCTCTCTGATCTCTGAGGCCAGGCTCTAATGTGATTTGAATCTACTTCTAACCCCTTCCAAGCACTGCCCTCCCGAATTCTCTGCTCCTCTCCCCACCCCACTGTTGGTCTGTGATTTCGAGGCAGGCGTGGCCCCCTGCAGCCTGGAATGAAGTCACTGGGGCTGTTTGGAGACCGGGGCTGTTTGGAGGTTAGACTGGGGTGGGAGTGGGGGATTGTGTTCTGGTATTGAGGTGCTGGTGTCTCTTGCTAGCGGGTGATGGGTGTGTCGTGCCCCTCCTGGCAGGACCCTGAGCCCCCTCCCTGGCCTTTCATCCTGTCTGTCTTCTTGAGTGTTTTCCCTTTCCACTCCTGCTCCTGTGGTACCCCTCTGGCCGGGACACCCAGGAGAATGTGTCAGGAGGAACTTCATCCACACAGGGCAGGGGCTCTGAGCGATGCTTTGGGTGGGGAGGGGCACCGCCGGGATGCTGGCTTCTCTCCAAATGGGCCTCTGAATCCAGGCCGGGCATTGCCTCTCTCCTTCCCAGGCGAGGGACTCCCTGGCGAGGAGGTGGTTTGGGACTGGGGGTGAGAGCAGATCGTAGGTCCCACCTGGCCTGGAGCTAGGCCGAGGGTGGGGAGGAGGAGGTGATGTCAGGAGCCTCAGGCCCACATCCTCCTCCCCCATCCTGTTCCCAGCCCCGTGCCTCCCCCATTCCGCTCTGGCTAGCAGTGGGTTGGGGAGTTGGGGGAAGGGTGGAGGCAGGGCTGTTTTTGTTTCTCCTAAGCCAGGAAGTTGTGTGGATGAGTCAAGGTTGAATGGAGAGCCCTAGGCAGTGTTGTCCCGCCCGCCCTGACTCACCTGTCCCCCAAATCCCCTGATGCCCTCACTGAAGGGTCAGGTGCCTGCTCCCTCCTGAGGGGTCAATCCCAGGTGCCCCAGGTGGCCTCATAGTACTGGCCCAGTGTTTGGAGAACAGGCTGGGGGCTGGGATTGCAGGTTCTCCCAGGTCTCAGCCCGGCCAAAGGGGTCTGCTGGGTTCCTGCGGAGAAGGTGCCAGCCCCACCTCCACCTCTAGTGCTCGTCTTTGCCTGCTGTCCTGTCTTTACTTGTCTTGAGTGTCTGTGCAGCCCCTATAAGTGATAAGACATGGGTGAGATTGGCCTTGCCTGGGGGTACCAGGTGCCCTCAGACAAATGGGTGGATAGGACAGGTGCACAGTAATGGATGATGCGGCCCCCGCACAGTGCTGAGAGAGAGTCATTCAGTCTGAATGTGGAAGGCCAGGAAGGCTTCTCAGAGGAGGTGGCTCTAGCACTGGGGCTTCAGGAAAGAGTTGGGCAGGAGGATGGCCTGGGGAGGGTACGCAAGGTTTGGGGGCTGCTGGGGCCAGGACCCCTACCCCAGGCCCTGCTCTCCCCCTTGCACTGCAGCGGAGCCCGCTCTCCTGCCTTTCTGACTCCCATCCTTCCCCACCTCCTTCTCCCCTCTGCATCACTCATCACTTGGGCAGGCCAAACCTCGCCAAGTATGAACAGGCTCTGGCAACTGGGTCTGGCAACAAAGTGCTGGAAAAACCCGGGGCACAGCCCAGCGCAGTGAGTGCCCCACACGTGGAGGCAAAGGAGTGGGCCTTTGCCCTGACCCTGAGGGCTGAGCCTGAGAGTCTGTGGGTTGTGAAATCTGTTGTGACCACCAGGCTTGTGGTGGGTGGACCCTCAGCTCTTCCTGTGAGCAACTTTGTGCTAAAGAAAGAGCACTGGAGCCGGGCGTGGTGGCTTACGCCTGTAATCCCAGCACTTTGGGAGGCCAAGGCGGGCGGATCAGGAGTTCAAGACCAGCCTGGCCAACATGGTGAAACCCTGTCTCTACTAAAAATACAAAAATTAGACGGGCGTGGTGGTGGGCACCTGTAATCCCAGTTACTCGGGAGGTTAAGGCACGAGAATTACTTGAACCCAGGAGGTGGAGGTTGCAGTGAGTGGAGATGACGCCGCTGCAAAAAAAAAAAAAAAAAAGGAGAAAAAAAACAGAAAGAGCACTGGACTTGGAGTCTAAACACAGGTTCTACTCCCGACTCACCTGTGACTGAGGGCAAGAATGCTTTCTTCTCAGAGACTCGGTGTCCCCTTTTGTAAAATGAGGCTCATTCTGTCCTTCCCCAAGGGCTTCAGGGAAACTCCAGTTAGAGCAGTGAGGTGCTAGGTAAACTCTGAGGGGCACCCTAACCGCCGTGTGAGGTCAGGAGGTCTGAATTCTCTCATCCCCTCTCCCCAGGACAAGGGCACACAACTGGTTCCGTTAAGCCCCTCTCTTGCTCAGACGCCATGGAGCTGGATCTGTCTCCACCTCATCTTAGCAGCTCTCCGGAAGACCTTTGCCCAGCCCCTGGGACCCCTCCTGGGACTCCCCGGCCCCCTGATACCCCTCTGCCTGAGGAGGTAAAGAGGTCCCAGCCTCTCCTCATCCCAACCACCGGCAGGTACGATGGGGCGTGGGGCTTGGGGGAGGTCAGTGCTGGATAATACACAGAGGCTTGCAGGCCACTGCTCCCTTCCCCACACCTCTCTCCCTTTTTCTTCTTGCCACAGGAAACTTCGAGAGGAGGAGAGGCGTGCCACCTCCCTCCCCTCTATCCCCAACCCCTTCCCTGAGCTCTGCAGTCCTCCCTCACAGAGCCCAATTCTCGGGGGCCCCTCCAGTGCAAGGGGGCTGCTCCCCCGCGATGCCAGCCGCCCCCATGTGAGTTGTCCCTCAGAAGGGAAGGGAGGGATGCACGGGTTCTGGGTCTGTGGGAGATACACAGCCGCTTCCATGGAGGCAGGGGATCTTGGTTAGGAGTCCCTGAGGGTCTAGCAGGTGCGGAAAGGGAATGAATCACCCTTTGCCTCCCCTCAAGTCGTGTGCAATTCCTGGGGCGCAGGTAGTAAAGGTGTACAGTGAGGATGGGGCCTGCAGGTCTGTGGAGGTGGCAGCAGGTGCCACAGCTCGCCACGTGTGTGAAATGCTGGTGCAGCGAGCTCACGCCTTGAGCGACGAGACCTGGGGGCTGGTGGAGTGCCACCCCCACCTAGCACTGGGTAAGTCAGGTGCATGGAACTATCCGGGCTGGGAGATGATGCCTTGCATCTTGGGCTAGGCATGTGGCTCATCCAGGAGATCTGGTTGATCTCCAATAACCCCTGCTTTTTGCCCTTGTCCCCAGAGCGGGGTTTGGAGGACCACGAGTCCGTGGTGGAAGTGCAGGCTGCCTGGCCCGTGGGCGGAGATAGCCGCTTCGTCTTCCGGAAAAACTTCGCCAAGTACGAACTGTTCAAGAGCTCCCCAGTGAGTGCATGAGGGCTGTCTGGGCGCTGGGATGCCCTGATCCTCAACCTGGATGCTGGAGCCCTGATCCCTGACACTTGTCTACCCACAGCACTCCCTGTTCCCAGAAAAAATGGTCTCCAGCTGTCTCGATGCACACACTGGTATATCCCATGAAGACCTCATCCAGGTGGGGGGACCCCCCATTTCACTGCAGATTCACGACTCCCCAGCATTGGCCAGTGCTTCTCCACCCTTAAGTCCTGTGCCTCCCCTCTATGTTGTAGAAAGAGCCAAATCACAGTCCTGTGTGACTGGGGACAGTCACTTTCCCTGCCTGAGCATCAGTTTCTTCTATTAAATGGGGGCGAGAAATGCATGTGGAGCATTTCCTTGTAAAAACCTGAGGGTGGGCTGGGCACGGTGGCTCATGCCTATAATCCCAGCACTTTGGGAGGCTGAGGCGGGAGGATTACTTAAGCCTAGAAGTTTGAGAGTTTGAGACCAGCCTGGGCAACATAATGAGACCTCGTCTCTCCAAAAAAAAAAAAAAAAAAAAAAAAAAGGCCAGGAATGGTGGCATGAGCCTGTAGTCCCAGGTGCTTGGGAGGCTGAGGTGGGAGGATCACTTGAGCTCGGGAGGTCGAGGTTGCAGTGAGCTGTGATCGTGCCACCGCACTAGCATGAGACCCTGTCTCAAAAAGAAAAAGAAAAAGAGAAACATCCTGGTGGTAGAGGGGAAGGGAGGAGGTCAGACCTGTCACTCTCCTCTGCTCTCCTCTGGCTCAGAACTTCCTGAATGCTGGCAGCTTTCCTGAGATCCAGGGCTTTCTGCAGCTGCGGGGTTCAGGACGGAAGCTTTGGAAACGCTTTTTCTGCTTCTTGCGCCGATCTGGCCTCTATTACTCCACCAAGGGCACCTCTAAGGTAAGGTCTTGAGGGTACCAGCCCCAGCCCCTCCAGTCCCTGGTCCTTTTAGAAGTTGCCCCTTCTCTGCTGGAACCTCTGAGCCCTTCTCCCCCTGGGCCCCCCAGGCCAGCCACCTCCAGTTTACCATCTCTCCCTACATCCTTGCCTAGCTCACCTGCCCAGGGAGGTAGCAGGAGAAAAGATGATCTTAGTTTAAGTCCTGGCTCTACTTCTATTTGCTGTGTGACCCTGGGTATTCCCCTGCCCCTCTCTGGTCCTGAAATCTCCCCACCTGGCTTGTGGGGGGAGGTAATAGTGGGCGGGATCTACCTGTACCAAGTCCTGCTCACTCATGCTGCTTAGGATCCGAGGCACCTGCAGTACGTGGCAGATGTGAACGAGTCCAACGTGTACGTGGTGACGCAGGGCCGCAAGCTCTACGGGATGCCCACTGACTTCGGTTTCTGTGTCAAGGTGAAGACCTGGCCAGGCCTGGCCCCTGGCCTGGGGAAGCAGGAACTGCTCAGGCCCCTGGATCCTGCCCGGGGCTTCTGAGCCCCAATTCAGACACCTAGACTCCTCTCCCTGCACCCTGGCCTGCTGGAAACTCCTGGATTCAGCTCTGCTATGTGGAGCAGGGGCAGACATGTGGTCCTAAAGGCAGATATGGGACCAGTCAATTTCCTCTCTCTGCAGCCCAACAAGCTTCGAAATGGCCACAAGGGGCTTCGGATCTTCTGCAGTGAAGATGAGCAGAGCCGCACCTGCTGGCTGGCTGCCTTCCGCCTCTTCAAGGTGAGACCCTGGGAGTGGCATGGGGGGCTGGCCTGGCCAGAGGGATCCCCAGCTCTGCCCTCAGGAAGTCTCAGGAATGAGGAGGGCATCACAGCCTTGCTCTCTGATAACCCCCAGTCCAAGCCTGAAGTTATAGGAAGTGCCCATCGAAGGCAGAAACACAGCCCTGGTCTGGGCAAGTCCTGGTCTGAGGGGGGCGTCACAGCCACGCCCCCAGGACCTCTCGACCTCAAGCTCTCTTTCTCTCCCCACCCCCAGTACGGGGTGCAGCTGTACAAGAATTACCAGCAGGCACAGTCTCGCCATCTGCATCCATCTTGTTTGGGCTCCCCACCCTTGGTGAGTGTGCCCAAGGGGATGGGAGGGTGGGTATGCAGGCCCTGTCTTACGGGTACCTGGGCCCTGTTCTGACCTCTCTCCTCTCCTTCCATCTCCAGAGAAGTGCCTCAGATAATACCCTGGTGGCCATGGACTTCTCTGGCCATGCTGGGCGTGTCATTGAGAACCCCCGGGAGGCTCTGAGTGTGGCCCTGGAGGAGGCCCAGGCCTGGAGGGTGAGGCCTGCTGTGTGTGTGTGTGTTTGTGCTGGGGACCCACTCTCTGGGTGGGATCCCTGAAATAGGAGGGAGGAAGAGAGGGCGGGGGGAGGCCCCTGGCTGGGAAGAAGTGCTCTACCTTCCTGAGGTGCTGGGTAATGCCCCCAAGCACGCCCCGACTCTCCCGTATCTCCCACTGCTCCACAGAAGAAGACAAACCACCGCCTCAGCCTGCCCATGCCAGCCTCCGGCACGAGCCTCAGTGCAGGTGGGTGACGGCCCCGAGTCCTGGGGCGGGGGCTGCCTCAACTTCTCTTTGTATTCCCAGTGGGGAGTAGATGGTATAGGGGTCCCCTCCCCAAAGTGACCGCCCATGTCCTTCCCCACCACAGCCATCCACCGCACCCAACTCTGGTTCCACGGGCGCATTTCCCGTGAGGAGAGCCAGCGGCTTATTGGACAGCAGGGCTTGGTAGACGGGTAAGGGGCAGGGCCGGGCAACAGACCCAGGGATAAGAGAGACTGGGGTCCAGGTGGCAGCCATGGTCCTTGGGTAGTAATGCTGCCCCATCTCCTGTCTTCTGGCAGCCTGTTCCTGGTCCGGGAGAGTCAGCGGAACCCCCAGGGCTTTGTCCTCTCTTTGTGCCACCTGCAGAAAGTGAAGCATTATCTCATCCTGCCGGTGAGCTTCCCTGCGTCCCCGGAGTCCTGCAATGAGACACAGGACTCCCAGCAACCTGTCCTCCTCACCAGGCCCCTCCAGAGGCTCCCTGGCCCCCAGTGCGTCTCCCTTTTCCCTGCACAAGAAGTGGGAGGCTGAGTGCGGTGGCTCACACCTGTAATCCCAGCACTTAGGACGGCCAAGGTGGGAGAATGGCTTGAGCCCAGGAGTTCGAGACCAGCCTGGGCAACACAGGGAGACCCCATCTCTACAAATAATTTAAAAATGAGCCAGGCATGGTGGTGCACACCTGTAGTCCAGCTACTCAGGAGGCTGAGGTGGGAGCATTGCTTGAGCCCAGAGGGTCAAGGCTGCAGTGAGCCATGGTGGCACCACCACACTCCAGCCTGGATGACACAGTGAGAAAACTGTCTCAAAAAAAAAAAAAGAAAGAAAAAGAAAAAGAAAAAAGAAAAGAATAAAAGGAAATGGTGGGGCCCTGGCCCAGGAGGGAGCTTCCCAAGCCTCGGGCCCCTCCCTGAACTTCCACCCCCTTTACTGTACCCCAGAGCGAGGAGGAGGGCCGCCTGTACTTCAGCATGGATGATGGCCAGACCCGCTTCACTGACCTGCTGCAGCTCGTGGAGTTCCACCAGCTGAACCGCGGCATCCTGCCGTGCTTGCTGCGCCATTGCTGCACGCGGGTGGCCCTCTGACCAGGCCGTGGACTGGCTCATGCCTCAGCCCGCCTTCAGGCTGCCCGCCGCCCCTCCACCCATCCAGTGGACTCTGGGGCGCGGCCACAGGGGACGGGATGAGGAGCGGGAGGGTTCCGCCACTCCAGTTTTCTCCTCTGCTTCTTTGCCTCCCTCAGATAGAAAACAGCCCCCACTCCAGTCCACTCCTGACCCCTCTCCTCAAGGGAAGGCCTTGGGTGGCCCCCTCTCCTTCTCCTAGCTCTGGAGGTGCTGCTCTAGGGCAGGGAATTATGGGAGAAGTGGGGGCAGCCCAGGCGGTTTCACGCCCCACACTTTGTACAGACCGAGAGGCCAGTTGATCTGCTCTGTTTTATACTAGTGACAATAAAGATTATTTTTTGATACA'"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resultados_diccionario_2[2886]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "894a6e6b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
